{
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "# Breakthrough flu flew through my door\n",
        "by Zhandos Ayupov, Polina Guseva"
      ],
      "metadata": {
        "id": "4Jt_tmpmP_y5"
      }
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "cPwkCuvYuZKw"
      },
      "source": [
        "## install and download"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "_KPOdLQDXIvb"
      },
      "source": [
        "### programmes"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "We need programmes for quality control (FastQC), aligning (bwa), and variant calling (samtools and varscan)"
      ],
      "metadata": {
        "id": "CtJ3TYhM0jjQ"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "2ggYowIIvANz",
        "outputId": "4459cd25-d8e6-4c0b-e3e0-8a23eca3cdd3"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree       \n",
            "Reading state information... Done\n",
            "The following package was automatically installed and is no longer required:\n",
            "  libnvidia-common-460\n",
            "Use 'apt autoremove' to remove it.\n",
            "The following additional packages will be installed:\n",
            "  libapache-pom-java libcommons-compress-java libcommons-jexl2-java\n",
            "  libcommons-logging-java libcommons-math3-java libcommons-parent-java\n",
            "  libhtsjdk-java libjbzip2-java libngs-java libngs-sdk-dev libngs-sdk1\n",
            "  libsnappy-java libsnappy-jni libxz-java\n",
            "Suggested packages:\n",
            "  libcommons-jexl2-java-doc libavalon-framework-java\n",
            "  libcommons-logging-java-doc libexcalibur-logkit-java liblog4j1.2-java\n",
            "  picard-tools\n",
            "The following NEW packages will be installed:\n",
            "  fastqc libapache-pom-java libcommons-compress-java libcommons-jexl2-java\n",
            "  libcommons-logging-java libcommons-math3-java libcommons-parent-java\n",
            "  libhtsjdk-java libjbzip2-java libngs-java libngs-sdk-dev libngs-sdk1\n",
            "  libsnappy-java libsnappy-jni libxz-java\n",
            "0 upgraded, 15 newly installed, 0 to remove and 5 not upgraded.\n",
            "Need to get 5,140 kB of archives.\n",
            "After this operation, 8,129 kB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu bionic/universe amd64 libcommons-math3-java all 3.6.1-3 [2,016 kB]\n",
            "Get:2 http://archive.ubuntu.com/ubuntu bionic/universe amd64 libjbzip2-java all 0.9.1-4 [42.6 kB]\n",
            "Get:3 http://archive.ubuntu.com/ubuntu bionic/universe amd64 libapache-pom-java all 18-1 [4,720 B]\n",
            "Get:4 http://archive.ubuntu.com/ubuntu bionic/universe amd64 libcommons-parent-java all 43-1 [10.8 kB]\n",
            "Get:5 http://archive.ubuntu.com/ubuntu bionic-updates/universe amd64 libcommons-compress-java all 1.18-1~18.04 [531 kB]\n",
            "Get:6 http://archive.ubuntu.com/ubuntu bionic/universe amd64 libcommons-logging-java all 1.2-2 [60.3 kB]\n",
            "Get:7 http://archive.ubuntu.com/ubuntu bionic/universe amd64 libcommons-jexl2-java all 2.1.1-4 [252 kB]\n",
            "Get:8 http://archive.ubuntu.com/ubuntu bionic/universe amd64 libngs-sdk1 amd64 1.3.0-3 [52.3 kB]\n",
            "Get:9 http://archive.ubuntu.com/ubuntu bionic/universe amd64 libngs-sdk-dev amd64 1.3.0-3 [92.8 kB]\n",
            "Get:10 http://archive.ubuntu.com/ubuntu bionic/universe amd64 libngs-java amd64 1.3.0-3 [76.4 kB]\n",
            "Get:11 http://archive.ubuntu.com/ubuntu bionic/universe amd64 libsnappy-jni amd64 1.1.4-1 [6,996 B]\n",
            "Get:12 http://archive.ubuntu.com/ubuntu bionic/universe amd64 libsnappy-java all 1.1.4-1 [67.2 kB]\n",
            "Get:13 http://archive.ubuntu.com/ubuntu bionic/universe amd64 libxz-java all 1.8-1 [134 kB]\n",
            "Get:14 http://archive.ubuntu.com/ubuntu bionic/universe amd64 libhtsjdk-java all 2.8.1+dfsg-2 [1,438 kB]\n",
            "Get:15 http://archive.ubuntu.com/ubuntu bionic/universe amd64 fastqc all 0.11.5+dfsg-6 [354 kB]\n",
            "Fetched 5,140 kB in 2s (2,679 kB/s)\n",
            "Selecting previously unselected package libcommons-math3-java.\n",
            "(Reading database ... 123942 files and directories currently installed.)\n",
            "Preparing to unpack .../00-libcommons-math3-java_3.6.1-3_all.deb ...\n",
            "Unpacking libcommons-math3-java (3.6.1-3) ...\n",
            "Selecting previously unselected package libjbzip2-java.\n",
            "Preparing to unpack .../01-libjbzip2-java_0.9.1-4_all.deb ...\n",
            "Unpacking libjbzip2-java (0.9.1-4) ...\n",
            "Selecting previously unselected package libapache-pom-java.\n",
            "Preparing to unpack .../02-libapache-pom-java_18-1_all.deb ...\n",
            "Unpacking libapache-pom-java (18-1) ...\n",
            "Selecting previously unselected package libcommons-parent-java.\n",
            "Preparing to unpack .../03-libcommons-parent-java_43-1_all.deb ...\n",
            "Unpacking libcommons-parent-java (43-1) ...\n",
            "Selecting previously unselected package libcommons-compress-java.\n",
            "Preparing to unpack .../04-libcommons-compress-java_1.18-1~18.04_all.deb ...\n",
            "Unpacking libcommons-compress-java (1.18-1~18.04) ...\n",
            "Selecting previously unselected package libcommons-logging-java.\n",
            "Preparing to unpack .../05-libcommons-logging-java_1.2-2_all.deb ...\n",
            "Unpacking libcommons-logging-java (1.2-2) ...\n",
            "Selecting previously unselected package libcommons-jexl2-java.\n",
            "Preparing to unpack .../06-libcommons-jexl2-java_2.1.1-4_all.deb ...\n",
            "Unpacking libcommons-jexl2-java (2.1.1-4) ...\n",
            "Selecting previously unselected package libngs-sdk1.\n",
            "Preparing to unpack .../07-libngs-sdk1_1.3.0-3_amd64.deb ...\n",
            "Unpacking libngs-sdk1 (1.3.0-3) ...\n",
            "Selecting previously unselected package libngs-sdk-dev.\n",
            "Preparing to unpack .../08-libngs-sdk-dev_1.3.0-3_amd64.deb ...\n",
            "Unpacking libngs-sdk-dev (1.3.0-3) ...\n",
            "Selecting previously unselected package libngs-java.\n",
            "Preparing to unpack .../09-libngs-java_1.3.0-3_amd64.deb ...\n",
            "Unpacking libngs-java (1.3.0-3) ...\n",
            "Selecting previously unselected package libsnappy-jni.\n",
            "Preparing to unpack .../10-libsnappy-jni_1.1.4-1_amd64.deb ...\n",
            "Unpacking libsnappy-jni (1.1.4-1) ...\n",
            "Selecting previously unselected package libsnappy-java.\n",
            "Preparing to unpack .../11-libsnappy-java_1.1.4-1_all.deb ...\n",
            "Unpacking libsnappy-java (1.1.4-1) ...\n",
            "Selecting previously unselected package libxz-java.\n",
            "Preparing to unpack .../12-libxz-java_1.8-1_all.deb ...\n",
            "Unpacking libxz-java (1.8-1) ...\n",
            "Selecting previously unselected package libhtsjdk-java.\n",
            "Preparing to unpack .../13-libhtsjdk-java_2.8.1+dfsg-2_all.deb ...\n",
            "Unpacking libhtsjdk-java (2.8.1+dfsg-2) ...\n",
            "Selecting previously unselected package fastqc.\n",
            "Preparing to unpack .../14-fastqc_0.11.5+dfsg-6_all.deb ...\n",
            "Unpacking fastqc (0.11.5+dfsg-6) ...\n",
            "Setting up libsnappy-jni (1.1.4-1) ...\n",
            "Setting up libcommons-math3-java (3.6.1-3) ...\n",
            "Setting up libjbzip2-java (0.9.1-4) ...\n",
            "Setting up libngs-sdk1 (1.3.0-3) ...\n",
            "Setting up libngs-sdk-dev (1.3.0-3) ...\n",
            "Setting up libxz-java (1.8-1) ...\n",
            "Setting up libsnappy-java (1.1.4-1) ...\n",
            "Setting up libapache-pom-java (18-1) ...\n",
            "Setting up libngs-java (1.3.0-3) ...\n",
            "Setting up libcommons-parent-java (43-1) ...\n",
            "Setting up libcommons-compress-java (1.18-1~18.04) ...\n",
            "Setting up libcommons-logging-java (1.2-2) ...\n",
            "Setting up libcommons-jexl2-java (2.1.1-4) ...\n",
            "Setting up libhtsjdk-java (2.8.1+dfsg-2) ...\n",
            "Setting up fastqc (0.11.5+dfsg-6) ...\n",
            "Processing triggers for man-db (2.8.3-2ubuntu0.1) ...\n",
            "Processing triggers for hicolor-icon-theme (0.17-2) ...\n",
            "Processing triggers for mime-support (3.60ubuntu1) ...\n",
            "Processing triggers for libc-bin (2.27-3ubuntu1.6) ...\n"
          ]
        }
      ],
      "source": [
        "!apt install fastqc"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "-lsQSVw1uTZf",
        "outputId": "9bb0e699-7c7d-4d2d-ba3d-8bf1a875d918"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree       \n",
            "Reading state information... Done\n",
            "The following package was automatically installed and is no longer required:\n",
            "  libnvidia-common-460\n",
            "Use 'apt autoremove' to remove it.\n",
            "Suggested packages:\n",
            "  samtools\n",
            "The following NEW packages will be installed:\n",
            "  bwa\n",
            "0 upgraded, 1 newly installed, 0 to remove and 5 not upgraded.\n",
            "Need to get 176 kB of archives.\n",
            "After this operation, 404 kB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu bionic-updates/universe amd64 bwa amd64 0.7.17-1ubuntu0.1 [176 kB]\n",
            "Fetched 176 kB in 1s (216 kB/s)\n",
            "Selecting previously unselected package bwa.\n",
            "(Reading database ... 124284 files and directories currently installed.)\n",
            "Preparing to unpack .../bwa_0.7.17-1ubuntu0.1_amd64.deb ...\n",
            "Unpacking bwa (0.7.17-1ubuntu0.1) ...\n",
            "Setting up bwa (0.7.17-1ubuntu0.1) ...\n",
            "Processing triggers for man-db (2.8.3-2ubuntu0.1) ...\n"
          ]
        }
      ],
      "source": [
        "!apt install bwa"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Please, take a note that the samtools version should be not older than 1.8. It is import for the further use of `mpileup` fuction"
      ],
      "metadata": {
        "id": "_qqSKNrksvBD"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "PBO0rFXsuiDW",
        "outputId": "decebc76-33a7-4a08-d12c-7906147111e5"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2022-11-14 18:16:25--  https://github.com/samtools/samtools/releases/download/1.12/samtools-1.12.tar.bz2\n",
            "Resolving github.com (github.com)... 192.30.255.113\n",
            "Connecting to github.com (github.com)|192.30.255.113|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/3666841/24fccc00-8707-11eb-9fab-b1ce3a33e4ea?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20221114%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20221114T181625Z&X-Amz-Expires=300&X-Amz-Signature=849318afcd0d254919d292deed1feb99995e264108e650abafb0cab3f2fe884f&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=3666841&response-content-disposition=attachment%3B%20filename%3Dsamtools-1.12.tar.bz2&response-content-type=application%2Foctet-stream [following]\n",
            "--2022-11-14 18:16:25--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/3666841/24fccc00-8707-11eb-9fab-b1ce3a33e4ea?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20221114%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20221114T181625Z&X-Amz-Expires=300&X-Amz-Signature=849318afcd0d254919d292deed1feb99995e264108e650abafb0cab3f2fe884f&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=3666841&response-content-disposition=attachment%3B%20filename%3Dsamtools-1.12.tar.bz2&response-content-type=application%2Foctet-stream\n",
            "Resolving objects.githubusercontent.com (objects.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...\n",
            "Connecting to objects.githubusercontent.com (objects.githubusercontent.com)|185.199.108.133|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 7521791 (7.2M) [application/octet-stream]\n",
            "Saving to: ‘samtools-1.12.tar.bz2’\n",
            "\n",
            "samtools-1.12.tar.b 100%[===================>]   7.17M  5.89MB/s    in 1.2s    \n",
            "\n",
            "2022-11-14 18:16:27 (5.89 MB/s) - ‘samtools-1.12.tar.bz2’ saved [7521791/7521791]\n",
            "\n",
            "make: Entering directory '/content/samtools-1.12'\n",
            "Makefile:111: config.mk: No such file or directory\n",
            "config.mk:34: htslib-1.12/htslib_static.mk: No such file or directory\n",
            "cd htslib-1.12 && make htslib.pc.tmp\n",
            "make[1]: Entering directory '/content/samtools-1.12/htslib-1.12'\n",
            "Makefile:118: htscodecs.mk: No such file or directory\n",
            "echo '# Default htscodecs.mk generated by Makefile' > htscodecs.mk\n",
            "echo 'include $(HTSPREFIX)htscodecs_bundled.mk' >> htscodecs.mk\n",
            "sed -e '/^static_libs=/s/@static_LIBS@/-lz -lm -lbz2 -llzma -lcurl/;s#@[^-][^@]*@##g' htslib.pc.in > htslib.pc.tmp\n",
            "make[1]: Leaving directory '/content/samtools-1.12/htslib-1.12'\n",
            "cd htslib-1.12 && make htslib_static.mk\n",
            "make[1]: Entering directory '/content/samtools-1.12/htslib-1.12'\n",
            "sed -n '/^static_libs=/s/[^=]*=/HTSLIB_static_LIBS = /p;/^static_ldflags=/s/[^=]*=/HTSLIB_static_LDFLAGS = /p' htslib.pc.tmp > htslib_static.mk\n",
            "make[1]: Leaving directory '/content/samtools-1.12/htslib-1.12'\n",
            "echo '/* Basic config.h generated by Makefile */' > config.h\n",
            "echo '#define HAVE_CURSES' >> config.h\n",
            "echo '#define HAVE_CURSES_H' >> config.h\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o bam_index.o bam_index.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o bam_plcmd.o bam_plcmd.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o sam_view.o sam_view.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o bam_fastq.o bam_fastq.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o bam_cat.o bam_cat.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o bam_md.o bam_md.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o bam_reheader.o bam_reheader.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o bam_sort.o bam_sort.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o bam_rmdup.o bam_rmdup.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o bam_rmdupse.o bam_rmdupse.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o bam_mate.o bam_mate.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o bam_stat.o bam_stat.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o bam_color.o bam_color.c\n",
            "echo '#define SAMTOOLS_VERSION \"1.12\"' > version.h\n",
            "echo '#define SAMTOOLS_CC \"gcc\"' >> version.h\n",
            "echo '#define SAMTOOLS_CPPFLAGS \"\"' >> version.h\n",
            "echo '#define SAMTOOLS_CFLAGS \"-g -Wall -O2\"' >> version.h\n",
            "echo '#define SAMTOOLS_LDFLAGS \"\"' >> version.h\n",
            "echo '#define SAMTOOLS_HTSDIR \"htslib-1.12\"' >> version.h\n",
            "echo '#define SAMTOOLS_LIBS \"\"' >> version.h\n",
            "echo '#define SAMTOOLS_CURSES_LIB \"-lcurses\"' >> version.h\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o bamtk.o bamtk.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o bam2bcf.o bam2bcf.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o bam2bcf_indel.o bam2bcf_indel.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o sample.o sample.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o cut_target.o cut_target.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o phase.o phase.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o bam2depth.o bam2depth.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o coverage.o coverage.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o padding.o padding.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o bedcov.o bedcov.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o bamshuf.o bamshuf.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o faidx.o faidx.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o dict.o dict.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o stats.o stats.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o stats_isize.o stats_isize.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o bam_flags.o bam_flags.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o bam_split.o bam_split.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o bam_tview.o bam_tview.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o bam_tview_curses.o bam_tview_curses.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o bam_tview_html.o bam_tview_html.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o bam_lpileup.o bam_lpileup.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o bam_quickcheck.o bam_quickcheck.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o bam_addrprg.o bam_addrprg.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o bam_markdup.o bam_markdup.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o tmp_file.o tmp_file.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o bam_ampliconclip.o bam_ampliconclip.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o amplicon_stats.o amplicon_stats.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o lz4/lz4.o lz4/lz4.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o bam_aux.o bam_aux.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o bam.o bam.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o sam.o sam.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o bam_plbuf.o bam_plbuf.c\n",
            "ar -csru libbam.a bam_aux.o bam.o sam.o bam_plbuf.o\n",
            "ar: `u' modifier ignored since `D' is the default (see `U')\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o sam_opts.o sam_opts.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o sam_utils.o sam_utils.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o bedidx.o bedidx.c\n",
            "ar -rcs libst.a sam_opts.o sam_utils.o bedidx.o\n",
            "cd htslib-1.12 && make config.h\n",
            "make[1]: Entering directory '/content/samtools-1.12/htslib-1.12'\n",
            "echo '/* Default config.h generated by Makefile */' > config.h\n",
            "echo '#ifndef _XOPEN_SOURCE' >> config.h\n",
            "echo '#define _XOPEN_SOURCE 600' >> config.h\n",
            "echo '#endif' >> config.h\n",
            "echo '#define HAVE_LIBBZ2 1' >> config.h\n",
            "echo '#define HAVE_LIBLZMA 1' >> config.h\n",
            "echo '#ifndef __APPLE__' >> config.h\n",
            "echo '#define HAVE_LZMA_H 1' >> config.h\n",
            "echo '#endif' >> config.h\n",
            "echo '#define HAVE_DRAND48 1' >> config.h\n",
            "echo '#define HAVE_LIBCURL 1' >> config.h\n",
            "make[1]: Leaving directory '/content/samtools-1.12/htslib-1.12'\n",
            "cd htslib-1.12 && make hts-object-files\n",
            "make[1]: Entering directory '/content/samtools-1.12/htslib-1.12'\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o kfunc.o kfunc.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o kstring.o kstring.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o bcf_sr_sort.o bcf_sr_sort.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o bgzf.o bgzf.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o errmod.o errmod.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o faidx.o faidx.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o header.o header.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o hfile.o hfile.c\n",
            "echo '#define HTS_VERSION_TEXT \"1.12\"' > version.h\n",
            "echo '#define HTS_CC \"gcc\"' > config_vars.h\n",
            "echo '#define HTS_CPPFLAGS \"\"' >> config_vars.h\n",
            "echo '#define HTS_CFLAGS \"-g -Wall -O2 -fvisibility=hidden\"' >> config_vars.h\n",
            "echo '#define HTS_LDFLAGS \"-fvisibility=hidden\"' >> config_vars.h\n",
            "echo '#define HTS_LIBS \"-lz -lm -lbz2 -llzma -lcurl\"' >> config_vars.h\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o hts.o hts.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o hts_expr.o hts_expr.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o hts_os.o hts_os.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o md5.o md5.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o multipart.o multipart.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o probaln.o probaln.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o realn.o realn.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o regidx.o regidx.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o region.o region.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o sam.o sam.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o synced_bcf_reader.o synced_bcf_reader.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o vcf_sweep.o vcf_sweep.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o tbx.o tbx.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o textutils.o textutils.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o thread_pool.o thread_pool.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o vcf.o vcf.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o vcfutils.o vcfutils.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o cram/cram_codecs.o cram/cram_codecs.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o cram/cram_decode.o cram/cram_decode.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o cram/cram_encode.o cram/cram_encode.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o cram/cram_external.o cram/cram_external.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o cram/cram_index.o cram/cram_index.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o cram/cram_io.o cram/cram_io.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o cram/cram_stats.o cram/cram_stats.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o cram/mFILE.o cram/mFILE.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o cram/open_trace_file.o cram/open_trace_file.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o cram/pooled_alloc.o cram/pooled_alloc.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o cram/string_alloc.o cram/string_alloc.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o htscodecs/htscodecs/arith_dynamic.o htscodecs/htscodecs/arith_dynamic.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o htscodecs/htscodecs/fqzcomp_qual.o htscodecs/htscodecs/fqzcomp_qual.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o htscodecs/htscodecs/htscodecs.o htscodecs/htscodecs/htscodecs.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o htscodecs/htscodecs/pack.o htscodecs/htscodecs/pack.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o htscodecs/htscodecs/rANS_static4x16pr.o htscodecs/htscodecs/rANS_static4x16pr.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o htscodecs/htscodecs/rANS_static.o htscodecs/htscodecs/rANS_static.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o htscodecs/htscodecs/rle.o htscodecs/htscodecs/rle.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o htscodecs/htscodecs/tokenise_name3.o htscodecs/htscodecs/tokenise_name3.c\n",
            "gcc -g -Wall -O2 -fvisibility=hidden -I.  -c -o hfile_libcurl.o hfile_libcurl.c\n",
            "touch hts-object-files\n",
            "make[1]: Leaving directory '/content/samtools-1.12/htslib-1.12'\n",
            "cd htslib-1.12 && make lib-static\n",
            "make[1]: Entering directory '/content/samtools-1.12/htslib-1.12'\n",
            "ar -rc libhts.a kfunc.o kstring.o bcf_sr_sort.o bgzf.o errmod.o faidx.o header.o hfile.o hts.o hts_expr.o hts_os.o md5.o multipart.o probaln.o realn.o regidx.o region.o sam.o synced_bcf_reader.o vcf_sweep.o tbx.o textutils.o thread_pool.o vcf.o vcfutils.o cram/cram_codecs.o cram/cram_decode.o cram/cram_encode.o cram/cram_external.o cram/cram_index.o cram/cram_io.o cram/cram_stats.o cram/mFILE.o cram/open_trace_file.o cram/pooled_alloc.o cram/string_alloc.o htscodecs/htscodecs/arith_dynamic.o htscodecs/htscodecs/fqzcomp_qual.o htscodecs/htscodecs/htscodecs.o htscodecs/htscodecs/pack.o htscodecs/htscodecs/rANS_static4x16pr.o htscodecs/htscodecs/rANS_static.o htscodecs/htscodecs/rle.o htscodecs/htscodecs/tokenise_name3.o hfile_libcurl.o\n",
            "ranlib libhts.a\n",
            "make[1]: Leaving directory '/content/samtools-1.12/htslib-1.12'\n",
            "gcc  -L./lz4  -o samtools bam_index.o bam_plcmd.o sam_view.o bam_fastq.o bam_cat.o bam_md.o bam_reheader.o bam_sort.o bam_rmdup.o bam_rmdupse.o bam_mate.o bam_stat.o bam_color.o bamtk.o bam2bcf.o bam2bcf_indel.o sample.o cut_target.o phase.o bam2depth.o coverage.o padding.o bedcov.o bamshuf.o faidx.o dict.o stats.o stats_isize.o bam_flags.o bam_split.o bam_tview.o bam_tview_curses.o bam_tview_html.o bam_lpileup.o bam_quickcheck.o bam_addrprg.o bam_markdup.o tmp_file.o bam_ampliconclip.o amplicon_stats.o ./lz4/lz4.o libbam.a libst.a htslib-1.12/libhts.a -lz -lm -lbz2 -llzma -lcurl -lcurses -lm -lz  -lpthread\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o misc/ace2sam.o misc/ace2sam.c\n",
            "gcc  -L./lz4  -o misc/ace2sam misc/ace2sam.o htslib-1.12/libhts.a -lz -lm -lbz2 -llzma -lcurl -lz \n",
            "gcc -g -Wall -O2 -DMAQ_LONGREADS -I. -Ihtslib-1.12 -I./lz4  -c -o misc/maq2sam-long.o misc/maq2sam.c\n",
            "gcc  -o misc/maq2sam-long misc/maq2sam-long.o -lz \n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o misc/maq2sam-short.o misc/maq2sam.c\n",
            "gcc  -o misc/maq2sam-short misc/maq2sam-short.o -lz \n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o misc/md5fa.o misc/md5fa.c\n",
            "gcc  -L./lz4  -o misc/md5fa misc/md5fa.o htslib-1.12/libhts.a -lz -lm -lbz2 -llzma -lcurl -lz \n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o misc/md5sum-lite.o misc/md5sum-lite.c\n",
            "gcc  -L./lz4  -o misc/md5sum-lite misc/md5sum-lite.o htslib-1.12/libhts.a -lz -lm -lbz2 -llzma -lcurl -lz \n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o misc/wgsim.o misc/wgsim.c\n",
            "gcc  -L./lz4  -o misc/wgsim misc/wgsim.o -lm htslib-1.12/libhts.a -lz -lm -lbz2 -llzma -lcurl -lz \n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o test/merge/test_bam_translate.o test/merge/test_bam_translate.c\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o test/test.o test/test.c\n",
            "gcc  -L./lz4  -o test/merge/test_bam_translate test/merge/test_bam_translate.o test/test.o libst.a htslib-1.12/libhts.a -lz -lm -lbz2 -llzma -lcurl -lz  -lpthread\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o test/merge/test_rtrans_build.o test/merge/test_rtrans_build.c\n",
            "gcc  -L./lz4  -o test/merge/test_rtrans_build test/merge/test_rtrans_build.o test/test.o libst.a htslib-1.12/libhts.a -lz -lm -lbz2 -llzma -lcurl -lz  -lpthread\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o test/merge/test_trans_tbl_init.o test/merge/test_trans_tbl_init.c\n",
            "gcc  -L./lz4  -o test/merge/test_trans_tbl_init test/merge/test_trans_tbl_init.o test/test.o libst.a htslib-1.12/libhts.a -lz -lm -lbz2 -llzma -lcurl -lz  -lpthread\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o test/split/test_count_rg.o test/split/test_count_rg.c\n",
            "gcc  -L./lz4  -o test/split/test_count_rg test/split/test_count_rg.o test/test.o libst.a htslib-1.12/libhts.a -lz -lm -lbz2 -llzma -lcurl -lz  -lpthread\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o test/split/test_expand_format_string.o test/split/test_expand_format_string.c\n",
            "gcc  -L./lz4  -o test/split/test_expand_format_string test/split/test_expand_format_string.o test/test.o libst.a htslib-1.12/libhts.a -lz -lm -lbz2 -llzma -lcurl -lz  -lpthread\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o test/split/test_filter_header_rg.o test/split/test_filter_header_rg.c\n",
            "gcc  -L./lz4  -o test/split/test_filter_header_rg test/split/test_filter_header_rg.o test/test.o libst.a htslib-1.12/libhts.a -lz -lm -lbz2 -llzma -lcurl -lz  -lpthread\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o test/split/test_parse_args.o test/split/test_parse_args.c\n",
            "gcc  -L./lz4  -o test/split/test_parse_args test/split/test_parse_args.o test/test.o libst.a htslib-1.12/libhts.a -lz -lm -lbz2 -llzma -lcurl -lz  -lpthread\n",
            "gcc -g -Wall -O2 -I. -Ihtslib-1.12 -I./lz4  -c -o test/vcf-miniview.o test/vcf-miniview.c\n",
            "gcc  -L./lz4  -o test/vcf-miniview test/vcf-miniview.o htslib-1.12/libhts.a -lz -lm -lbz2 -llzma -lcurl -lz  -lpthread\n",
            "make: Leaving directory '/content/samtools-1.12'\n",
            "make: Entering directory '/content/samtools-1.12'\n",
            "mkdir -p -m 755 /usr/local/bin /usr/local/bin /usr/local/share/man/man1\n",
            "install -p samtools /usr/local/bin\n",
            "install -p misc/ace2sam misc/maq2sam-long misc/maq2sam-short misc/md5fa misc/md5sum-lite misc/wgsim /usr/local/bin\n",
            "install -p misc/blast2sam.pl misc/bowtie2sam.pl misc/export2sam.pl misc/fasta-sanitize.pl misc/interpolate_sam.pl misc/novo2sam.pl misc/plot-ampliconstats misc/plot-bamstats misc/psl2sam.pl misc/sam2vcf.pl misc/samtools.pl misc/seq_cache_populate.pl misc/soap2sam.pl misc/wgsim_eval.pl misc/zoom2sam.pl /usr/local/bin\n",
            "install -p -m 644 doc/samtools*.1 misc/wgsim.1 /usr/local/share/man/man1\n",
            "make: Leaving directory '/content/samtools-1.12'\n"
          ]
        }
      ],
      "source": [
        "!wget https://github.com/samtools/samtools/releases/download/1.12/samtools-1.12.tar.bz2\n",
        "!tar -xf samtools-1.12.tar.bz2\n",
        "!sudo make -C samtools-1.12/\n",
        "!sudo make install -C samtools-1.12/"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "KD3bmlsK78q_",
        "outputId": "699228f8-d5da-46f7-9f61-2fab3e6a1433"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree       \n",
            "Reading state information... Done\n",
            "The following package was automatically installed and is no longer required:\n",
            "  libnvidia-common-460\n",
            "Use 'apt autoremove' to remove it.\n",
            "The following NEW packages will be installed:\n",
            "  varscan\n",
            "0 upgraded, 1 newly installed, 0 to remove and 5 not upgraded.\n",
            "Need to get 138 kB of archives.\n",
            "After this operation, 166 kB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu bionic/multiverse amd64 varscan amd64 2.4.3+dfsg-1 [138 kB]\n",
            "Fetched 138 kB in 0s (371 kB/s)\n",
            "Selecting previously unselected package varscan.\n",
            "(Reading database ... 124301 files and directories currently installed.)\n",
            "Preparing to unpack .../varscan_2.4.3+dfsg-1_amd64.deb ...\n",
            "Unpacking varscan (2.4.3+dfsg-1) ...\n",
            "Setting up varscan (2.4.3+dfsg-1) ...\n",
            "Processing triggers for man-db (2.8.3-2ubuntu0.1) ...\n"
          ]
        }
      ],
      "source": [
        "!apt install -y varscan"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!apt install bedtools"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "IyJaboifQegg",
        "outputId": "3e8c7698-0784-4d15-938d-dfbfdda06fad"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree       \n",
            "Reading state information... Done\n",
            "The following package was automatically installed and is no longer required:\n",
            "  libnvidia-common-460\n",
            "Use 'apt autoremove' to remove it.\n",
            "The following NEW packages will be installed:\n",
            "  bedtools\n",
            "0 upgraded, 1 newly installed, 0 to remove and 5 not upgraded.\n",
            "Need to get 577 kB of archives.\n",
            "After this operation, 2,040 kB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu bionic/universe amd64 bedtools amd64 2.26.0+dfsg-5 [577 kB]\n",
            "Fetched 577 kB in 1s (507 kB/s)\n",
            "Selecting previously unselected package bedtools.\n",
            "(Reading database ... 124311 files and directories currently installed.)\n",
            "Preparing to unpack .../bedtools_2.26.0+dfsg-5_amd64.deb ...\n",
            "Unpacking bedtools (2.26.0+dfsg-5) ...\n",
            "Setting up bedtools (2.26.0+dfsg-5) ...\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "JNPwpi3SXOEJ"
      },
      "source": [
        "### data"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Here we upload:\n",
        "\n",
        "*   our test reads (sequences from out sick roommate) conoformed to be from the A/Hong Kong/4801/2014 (H3N2) strain;\n",
        "*   as for the reference, KF848938.1 Influenza A virus (A/USA/RVD1_H3/2011(H3N2)) segment 4 hemagglutinin (HA) gene;\n",
        "*   and controls as the isogenic (100% pure) sample of the standard (reference) H3N2 influenza virus run by the same sequenator.\n",
        "\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "JuyTaR7O01PS"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "iZAsuGbsvb33",
        "outputId": "e3853617-3757-4d05-932d-f1deeb7a3c02"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2022-11-14 18:17:46--  http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR170/001/SRR1705851/SRR1705851.fastq.gz\n",
            "Resolving ftp.sra.ebi.ac.uk (ftp.sra.ebi.ac.uk)... 193.62.193.138\n",
            "Connecting to ftp.sra.ebi.ac.uk (ftp.sra.ebi.ac.uk)|193.62.193.138|:80... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 19217176 (18M) [application/x-gzip]\n",
            "Saving to: ‘SRR1705851.fastq.gz’\n",
            "\n",
            "SRR1705851.fastq.gz 100%[===================>]  18.33M   483KB/s    in 39s     \n",
            "\n",
            "2022-11-14 18:18:26 (479 KB/s) - ‘SRR1705851.fastq.gz’ saved [19217176/19217176]\n",
            "\n"
          ]
        }
      ],
      "source": [
        "# our reads\n",
        "!wget http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR170/001/SRR1705851/SRR1705851.fastq.gz"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "PTCzYjK2xTP7"
      },
      "outputs": [],
      "source": [
        "!mv /content/SRR1705851.fastq.gz roommate.fastq.gz"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "N2pPyTbPw8hs",
        "outputId": "ef1a522a-d233-440c-f053-7ad79a407dae"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Downloading...\n",
            "From: https://drive.google.com/uc?id=1bDDN2R43l8SjgcFRv5QkaFduysTH2MwI\n",
            "To: /content/sequence.fasta\n",
            "\r  0% 0.00/1.79k [00:00<?, ?B/s]\r100% 1.79k/1.79k [00:00<00:00, 2.73MB/s]\n"
          ]
        }
      ],
      "source": [
        "# our reference\n",
        "!gdown 1bDDN2R43l8SjgcFRv5QkaFduysTH2MwI"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "rKpPF0Ag5C3z"
      },
      "outputs": [],
      "source": [
        "!mv /content/sequence.fasta reference.fasta"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "jL9TunZeT7Tw",
        "outputId": "7acdb0c2-eaa5-4181-c78b-26645f6c12d6"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2022-11-14 18:18:29--  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR170/008/SRR1705858/SRR1705858.fastq.gz\n",
            "           => ‘SRR1705858.fastq.gz’\n",
            "Resolving ftp.sra.ebi.ac.uk (ftp.sra.ebi.ac.uk)... 193.62.193.138\n",
            "Connecting to ftp.sra.ebi.ac.uk (ftp.sra.ebi.ac.uk)|193.62.193.138|:21... connected.\n",
            "Logging in as anonymous ... Logged in!\n",
            "==> SYST ... done.    ==> PWD ... done.\n",
            "==> TYPE I ... done.  ==> CWD (1) /vol1/fastq/SRR170/008/SRR1705858 ... done.\n",
            "==> SIZE SRR1705858.fastq.gz ... 14782243\n",
            "==> PASV ... done.    ==> RETR SRR1705858.fastq.gz ... done.\n",
            "Length: 14782243 (14M) (unauthoritative)\n",
            "\n",
            "SRR1705858.fastq.gz 100%[===================>]  14.10M  8.16MB/s    in 1.7s    \n",
            "\n",
            "2022-11-14 18:18:34 (8.16 MB/s) - ‘SRR1705858.fastq.gz’ saved [14782243]\n",
            "\n",
            "--2022-11-14 18:18:34--  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR170/009/SRR1705859/SRR1705859.fastq.gz\n",
            "           => ‘SRR1705859.fastq.gz’\n",
            "Resolving ftp.sra.ebi.ac.uk (ftp.sra.ebi.ac.uk)... 193.62.193.138\n",
            "Connecting to ftp.sra.ebi.ac.uk (ftp.sra.ebi.ac.uk)|193.62.193.138|:21... connected.\n",
            "Logging in as anonymous ... Logged in!\n",
            "==> SYST ... done.    ==> PWD ... done.\n",
            "==> TYPE I ... done.  ==> CWD (1) /vol1/fastq/SRR170/009/SRR1705859 ... done.\n",
            "==> SIZE SRR1705859.fastq.gz ... 13581641\n",
            "==> PASV ... done.    ==> RETR SRR1705859.fastq.gz ... done.\n",
            "Length: 13581641 (13M) (unauthoritative)\n",
            "\n",
            "SRR1705859.fastq.gz 100%[===================>]  12.95M  7.27MB/s    in 1.8s    \n",
            "\n",
            "2022-11-14 18:18:38 (7.27 MB/s) - ‘SRR1705859.fastq.gz’ saved [13581641]\n",
            "\n",
            "--2022-11-14 18:18:38--  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR170/000/SRR1705860/SRR1705860.fastq.gz\n",
            "           => ‘SRR1705860.fastq.gz’\n",
            "Resolving ftp.sra.ebi.ac.uk (ftp.sra.ebi.ac.uk)... 193.62.193.138\n",
            "Connecting to ftp.sra.ebi.ac.uk (ftp.sra.ebi.ac.uk)|193.62.193.138|:21... connected.\n",
            "Logging in as anonymous ... Logged in!\n",
            "==> SYST ... done.    ==> PWD ... done.\n",
            "==> TYPE I ... done.  ==> CWD (1) /vol1/fastq/SRR170/000/SRR1705860 ... done.\n",
            "==> SIZE SRR1705860.fastq.gz ... 14372113\n",
            "==> PASV ... done.    ==> RETR SRR1705860.fastq.gz ... done.\n",
            "Length: 14372113 (14M) (unauthoritative)\n",
            "\n",
            "SRR1705860.fastq.gz 100%[===================>]  13.71M  3.64MB/s    in 3.8s    \n",
            "\n",
            "2022-11-14 18:18:45 (3.64 MB/s) - ‘SRR1705860.fastq.gz’ saved [14372113]\n",
            "\n"
          ]
        }
      ],
      "source": [
        "# three controls (from sequencing of isogenic reference samples)\n",
        "!wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR170/008/SRR1705858/SRR1705858.fastq.gz\n",
        "!wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR170/009/SRR1705859/SRR1705859.fastq.gz\n",
        "!wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR170/000/SRR1705860/SRR1705860.fastq.gz"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "n0VQn_UrWubu"
      },
      "outputs": [],
      "source": [
        "!mv /content/SRR1705858.fastq.gz control1.fastq.gz\n",
        "!mv /content/SRR1705859.fastq.gz control2.fastq.gz\n",
        "!mv /content/SRR1705860.fastq.gz control3.fastq.gz"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "S_ezD11FudN0"
      },
      "source": [
        "## Inspect the data from your roommate"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Let's have a look to our test reads"
      ],
      "metadata": {
        "id": "4S0opMAKu8Gm"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "a_w5zK_tvX02",
        "outputId": "813c7f83-d111-4374-aaa7-042c8f1bc6e7"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "@SRR1705851.1 1/1\n",
            "TTCGTGATTGTTTTCACTATCGTTCCGTTTGGCACTGCATGGTGCCCAAGGCACAGCGTTGCCGTGCTGTTGTCATTTCCAGGAAGTTTTTGAGCGAAAACCAGACATAGAATGTAGCTCAAAGCAATGATAGTCTTCATGGTTAATAG\n",
            "+\n",
            ",<==<<<<A@@@@@@@EEE;CEE+AC>EC;>EFFDC@=A@AE999DDD>>@E777EE75C>EF>EDEEFFFF--AE>EDEEEED=C-58AE=<D=<<DD=D9CDD@EEDED@DEDDE*9;@DDED@@@7@E*;*888@*8;@8@;;@@E\n",
            "@SRR1705851.2 2/1\n",
            "NATTAACCATGAAGACTATCATTGCTTTGAGCTACATTCTATGTCTGGTTTTCGCTCAAAAACTTCCTGGAAATGACAACAGCACGGCAACGCTGTGCCTTGGGCACCATGCAGTGCCAAACGGAACGATAGTGAAAACAATCACGAATGA\n",
            "+\n",
            "#5<???BBEEEDEDDDGGGGGGIIIIIIIIIIIIIIIIIIIIIHIIIIFHHIIHHHHHIIIIHIIIIIIIHIIIIIIIIIIIIIIHHHHHHHHHHEHHHHHFFHHHHHHFFHHGFGGGGGGGGGGGGGEEEGCEEGGGGGEEGGGGCGEGG\n",
            "@SRR1705851.3 3/1\n",
            "GTTTGGCACTGCATGGTGCCCAAGGCACAGCGTTGCCGTGCTGTTGTCATTTCCAGGAAGTTTTTGAGCGAAAACCAGACATAGAATGTAGCTCAAAGCAATGATAGTCTTCATGGTTAATAG\n"
          ]
        }
      ],
      "source": [
        "!zcat roommate.fastq.gz | head"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Let's count amount of reads and calculate an average coverage"
      ],
      "metadata": {
        "id": "rWcTmd0z2wg5"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "sRvsCMmOwNl0",
        "outputId": "e10da4f2-a1d6-4f14-a598-ffc2d2be5a31"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "65482 roommate.fastq.gz\n"
          ]
        }
      ],
      "source": [
        "! wc -l roommate.fastq.gz"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "2DHUBNgJz2O6",
        "outputId": "73e4627f-1a92-4ce4-fa64-50f9bb4a8bdd"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Total number of reads: 16,370\n"
          ]
        }
      ],
      "source": [
        "print(f\"Total number of reads: {int(65482 / 4):,}\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "dEf1CGna6GqI"
      },
      "outputs": [],
      "source": [
        "!gunzip roommate.fastq.gz"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "with open('/content/roommate.fastq') as reads_fastq:\n",
        "  list_of_reads = reads_fastq.readlines()\n",
        "\n",
        "total_length = [len(i.strip()) for i in list_of_reads[1::4]]\n",
        "print(f'The total sum length of all reads is {sum(total_length):,} bp')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "hhdz3pcUBlPx",
        "outputId": "862d8a20-fa4d-4425-adec-42de4499c42c"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "The total sum length of all reads is 52,717,864 bp\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Let's have a look at a lenght of our reference"
      ],
      "metadata": {
        "id": "-bIRPsiEEJD-"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!grep -v \">\" reference.fasta | grep -E -o \"G|C|T|A|N\" | wc -l"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "DngzyCzKEIil",
        "outputId": "96a8ffd7-9c5e-4b94-c825-1c0e66f7a8ae"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "1665\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "print(f'Therefore, the average coverage of our sequence reads is ~{52717864 // 1665:,} reads/nucleotide.')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "XoAGPe_YDx7O",
        "outputId": "fa5fa86f-0c53-4d65-804f-d9df8e4c926a"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Therefore, the average coverage of our sequence reads is ~31,662 reads/nucleotide.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Let's inspect our data quality"
      ],
      "metadata": {
        "id": "aVD5wEc1vNer"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "zkgn4rIR0GG6"
      },
      "outputs": [],
      "source": [
        "!mkdir fastqc_reports"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "3U2qAMnp0P2H",
        "outputId": "8cb0c8e8-4029-447a-d1af-153fcc44f932"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Started analysis of roommate.fastq\n",
            "Approx 5% complete for roommate.fastq\n",
            "Approx 10% complete for roommate.fastq\n",
            "Approx 15% complete for roommate.fastq\n",
            "Approx 20% complete for roommate.fastq\n",
            "Approx 25% complete for roommate.fastq\n",
            "Approx 30% complete for roommate.fastq\n",
            "Approx 35% complete for roommate.fastq\n",
            "Approx 40% complete for roommate.fastq\n",
            "Approx 45% complete for roommate.fastq\n",
            "Approx 50% complete for roommate.fastq\n",
            "Approx 55% complete for roommate.fastq\n",
            "Approx 60% complete for roommate.fastq\n",
            "Approx 65% complete for roommate.fastq\n",
            "Approx 70% complete for roommate.fastq\n",
            "Approx 75% complete for roommate.fastq\n",
            "Approx 80% complete for roommate.fastq\n",
            "Approx 85% complete for roommate.fastq\n",
            "Approx 90% complete for roommate.fastq\n",
            "Approx 95% complete for roommate.fastq\n",
            "Analysis complete for roommate.fastq\n"
          ]
        }
      ],
      "source": [
        "!fastqc -t 4 -o fastqc_reports/ roommate.fastq"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "e2Slf_I82Sab"
      },
      "source": [
        "The FastQC report warns about: \n",
        "*  Per base sequence content\n",
        "*  Sequence Duplication Levels\n",
        "\n",
        "All of these comments could be ignored, since we sequenced just one short-length fragment with a redundant coverage which caused a high level of duplications. Also, RNA libraries are known to have unequal random primer hybridisation leading to base disbalance. "
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "WhZjddiY7TjA"
      },
      "source": [
        "## Align your roommate’s data to the reference sequence"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Let's index the reference for the further alignment by the Burrows-Wheeler algorithm"
      ],
      "metadata": {
        "id": "RXZuYIuLxE7F"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "dhHDv-gb2dh-",
        "outputId": "1f4415f1-65ad-46fb-9f0b-64cc5d67254d"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[bwa_index] Pack FASTA... 0.00 sec\n",
            "[bwa_index] Construct BWT for the packed sequence...\n",
            "[bwa_index] 0.00 seconds elapse.\n",
            "[bwa_index] Update BWT... 0.00 sec\n",
            "[bwa_index] Pack forward-only FASTA... 0.00 sec\n",
            "[bwa_index] Construct SA from BWT and Occ... 0.00 sec\n",
            "[main] Version: 0.7.17-r1188\n",
            "[main] CMD: bwa index reference.fasta\n",
            "[main] Real time: 0.044 sec; CPU: 0.017 sec\n"
          ]
        }
      ],
      "source": [
        "!bwa index reference.fasta"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "To avoid unnesesary interim files, the aligned *.sam file directly goes to the next step. '`view`' views and converts SAM/BAM/CRAM files, '-S' was important for previous samtools versions to correct format, now it's done automatically, so just legacy. '-b' is for output in the BAM format. The BAM format is more compact. Sorting is helpful with further processing"
      ],
      "metadata": {
        "id": "q3fcbA2axYJX"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Lbm_QUFE42qB",
        "outputId": "bb1ca70b-089a-475d-8be0-65daa135df75"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
            "[M::process] read 68388 sequences (10000129 bp)...\n",
            "[M::process] read 67628 sequences (10000233 bp)...\n",
            "[M::mem_process_seqs] Processed 68388 reads in 2.493 CPU sec, 2.443 real sec\n",
            "[M::process] read 67698 sequences (10000046 bp)...\n",
            "[M::mem_process_seqs] Processed 67628 reads in 2.439 CPU sec, 2.359 real sec\n",
            "[M::process] read 67652 sequences (10000169 bp)...\n",
            "[M::mem_process_seqs] Processed 67698 reads in 2.985 CPU sec, 3.021 real sec\n",
            "[M::process] read 68072 sequences (10000295 bp)...\n",
            "[M::mem_process_seqs] Processed 67652 reads in 3.068 CPU sec, 2.983 real sec\n",
            "[M::process] read 18827 sequences (2716992 bp)...\n",
            "[M::mem_process_seqs] Processed 68072 reads in 2.570 CPU sec, 2.657 real sec\n",
            "[M::mem_process_seqs] Processed 18827 reads in 1.551 CPU sec, 1.526 real sec\n",
            "[main] Version: 0.7.17-r1188\n",
            "[main] CMD: bwa mem reference.fasta roommate.fastq\n",
            "[main] Real time: 15.426 sec; CPU: 15.210 sec\n"
          ]
        }
      ],
      "source": [
        "!bwa mem reference.fasta roommate.fastq | samtools view -S -b - | samtools sort -o roommate_sorted.bam -"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Let's have a quick look at the alignment statistics"
      ],
      "metadata": {
        "id": "0ZdAGcDCydrA"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "YidL3bap6vLG",
        "outputId": "8a0a404a-6fe0-4fb9-a0fe-0fc7973d5f0b"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "361349 + 0 in total (QC-passed reads + QC-failed reads)\n",
            "0 + 0 secondary\n",
            "3084 + 0 supplementary\n",
            "0 + 0 duplicates\n",
            "361116 + 0 mapped (99.94% : N/A)\n",
            "0 + 0 paired in sequencing\n",
            "0 + 0 read1\n",
            "0 + 0 read2\n",
            "0 + 0 properly paired (N/A : N/A)\n",
            "0 + 0 with itself and mate mapped\n",
            "0 + 0 singletons (N/A : N/A)\n",
            "0 + 0 with mate mapped to a different chr\n",
            "0 + 0 with mate mapped to a different chr (mapQ>=5)\n"
          ]
        }
      ],
      "source": [
        "!samtools flagstat roommate_sorted.bam"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "For quick searching through and visualisation in IGV let's add indexes to the sorted *.bam version and the reference"
      ],
      "metadata": {
        "id": "YU30k-W3ywii"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "YMXXg_d865zu"
      },
      "outputs": [],
      "source": [
        "!samtools index roommate_sorted.bam"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "zy6gwod-7CGu"
      },
      "outputs": [],
      "source": [
        "!samtools faidx reference.fasta"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ryFvNRbt7age"
      },
      "source": [
        "## Look for common variants with VarScan"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "bOA8mM019Dma"
      },
      "source": [
        "To save computing power, the default behavior of samtools is to stop piling up the base calls at each position when it gets to 8000 calls. However, our variants may be quite rare and the average coverage x4 deeper. So, to keep all possible variants, set that depth limit to the max coverage with the `-d 0` flag.\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "E80ifeWL9MXl",
        "outputId": "12b3c4a6-7bea-44fe-e8b6-80a29219ace5"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "Usage: samtools mpileup [options] in1.bam [in2.bam [...]]\n",
            "\n",
            "Input options:\n",
            "  -6, --illumina1.3+      quality is in the Illumina-1.3+ encoding\n",
            "  -A, --count-orphans     do not discard anomalous read pairs\n",
            "  -b, --bam-list FILE     list of input BAM filenames, one per line\n",
            "  -B, --no-BAQ            disable BAQ (per-Base Alignment Quality)\n",
            "  -C, --adjust-MQ INT     adjust mapping quality; recommended:50, disable:0 [0]\n",
            "  -d, --max-depth INT     max per-file depth; avoids excessive memory usage [8000]\n",
            "  -E, --redo-BAQ          recalculate BAQ on the fly, ignore existing BQs\n",
            "  -f, --fasta-ref FILE    faidx indexed reference sequence file\n",
            "  -G, --exclude-RG FILE   exclude read groups listed in FILE\n",
            "  -l, --positions FILE    skip unlisted positions (chr pos) or regions (BED)\n",
            "  -q, --min-MQ INT        skip alignments with mapQ smaller than INT [0]\n",
            "  -Q, --min-BQ INT        skip bases with baseQ/BAQ smaller than INT [13]\n",
            "  -r, --region REG        region in which pileup is generated\n",
            "  -R, --ignore-RG         ignore RG tags (one BAM = one sample)\n",
            "  --rf, --incl-flags STR|INT  required flags: skip reads with mask bits unset []\n",
            "  --ff, --excl-flags STR|INT  filter flags: skip reads with mask bits set\n",
            "                                            [UNMAP,SECONDARY,QCFAIL,DUP]\n",
            "  -x, --ignore-overlaps   disable read-pair overlap detection\n",
            "  -X, --customized-index  use customized index files\n",
            "\n",
            "Output options:\n",
            "  -o, --output FILE       write output to FILE [standard output]\n",
            "  -O, --output-BP         output base positions on reads\n",
            "  -s, --output-MQ         output mapping quality\n",
            "      --output-QNAME      output read names\n",
            "      --output-extra STR  output extra read fields and read tag values\n",
            "      --output-sep CHAR   set the separator character for tag lists [,]\n",
            "      --output-empty CHAR set the no value character for tag lists [*]\n",
            "      --reverse-del       use '#' character for deletions on the reverse strand\n",
            "  -a                      output all positions (including zero depth)\n",
            "  -a -a (or -aa)          output absolutely all positions, including unused ref. sequences\n",
            "\n",
            "Generic options:\n",
            "      --input-fmt-option OPT[=VAL]\n",
            "               Specify a single input file format option in the form\n",
            "               of OPTION or OPTION=VALUE\n",
            "      --reference FILE\n",
            "               Reference sequence FASTA FILE [null]\n",
            "      --verbosity INT\n",
            "               Set level of verbosity\n",
            "\n",
            "Note that using \"samtools mpileup\" to generate BCF or VCF files is now\n",
            "deprecated.  To output these formats, please use \"bcftools mpileup\" instead.\n"
          ]
        }
      ],
      "source": [
        "!samtools mpileup"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "hGBT8N9O7KYs",
        "outputId": "843346b9-3258-4670-a567-5ef0d7c224e8"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[mpileup] 1 samples in 1 input files\n",
            "[mpileup] Max depth set to maximum value (2147483647)\n"
          ]
        }
      ],
      "source": [
        "!samtools mpileup -d 0 -f reference.fasta -o roommate_aligned.mpileup roommate_sorted.bam"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Due to the influeza virus being one-strand, we can turn off this basic preventive setting"
      ],
      "metadata": {
        "id": "dJFdmA-Gzoch"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!varscan mpileup2snp --help"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ouzc0Gj0sehX",
        "outputId": "a34e837d-b7e2-42ba-d111-168734057f01"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Only SNPs will be reported\n",
            "Warning: No p-value threshold provided, so p-values will not be calculated\n",
            "Min coverage:\t8\n",
            "Min reads2:\t2\n",
            "Min var freq:\t0.2\n",
            "Min avg qual:\t15\n",
            "P-value thresh:\t0.01\n",
            "USAGE: varscan mpileup2cns [pileup file] OPTIONS\n",
            "\tmpileup file - The SAMtools mpileup file\n",
            "\n",
            "\tOPTIONS:\n",
            "\t--min-coverage\tMinimum read depth at a position to make a call [8]\n",
            "\t--min-reads2\tMinimum supporting reads at a position to call variants [2]\n",
            "\t--min-avg-qual\tMinimum base quality at a position to count a read [15]\n",
            "\t--min-var-freq\tMinimum variant allele frequency threshold [0.01]\n",
            "\t--min-freq-for-hom\tMinimum frequency to call homozygote [0.75]\n",
            "\t--p-value\tDefault p-value threshold for calling variants [99e-02]\n",
            "\t--strand-filter\tIgnore variants with >90% support on one strand [1]\n",
            "\t--output-vcf\tIf set to 1, outputs in VCF format\n",
            "\t--vcf-sample-list\tFor VCF output, a list of sample names in order, one per line\n",
            "\t--variants\tReport only variant (SNP/indel) positions [0]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Let's have a look how many really established mutations there are (at least 95% frequency)"
      ],
      "metadata": {
        "id": "20KQmWUqz9IZ"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "YBtwflCt72TW",
        "outputId": "bf48feaf-40c1-4b0a-d16d-4fcd5895667a"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Only SNPs will be reported\n",
            "Warning: No p-value threshold provided, so p-values will not be calculated\n",
            "Min coverage:\t8\n",
            "Min reads2:\t2\n",
            "Min var freq:\t0.95\n",
            "Min avg qual:\t15\n",
            "P-value thresh:\t0.01\n",
            "Reading input from roommate_aligned.mpileup\n",
            "1665 bases in pileup file\n",
            "5 variant positions (5 SNP, 0 indel)\n",
            "0 were failed by the strand-filter\n",
            "5 variant positions reported (5 SNP, 0 indel)\n"
          ]
        }
      ],
      "source": [
        "!varscan mpileup2snp roommate_aligned.mpileup --min-var-freq 0.95 --strand-filter 0 --variants --output-vcf 1 > varscan_results_095.vcf"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ZfWHVXaF9j0Y"
      },
      "source": [
        "Discovered: 5 SNPs, if we DO NOT ignore variants with >90% support on one strand (```--strand-filter 0```), since our virus is one-stranded.\n",
        "\n",
        "\n",
        "Here they are: listed from left to right as my favourite BTS"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "7LjCgoN_9k7F",
        "outputId": "c390c757-1293-40e0-b9a2-2dec3d5fb562"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "KF848938.1 72 A G\n",
            "KF848938.1 117 C T\n",
            "KF848938.1 774 T C\n",
            "KF848938.1 999 C T\n",
            "KF848938.1 1260 A C\n"
          ]
        }
      ],
      "source": [
        "!cat varscan_results_095.vcf | awk 'NR>24 {print $1, $2, $4, $5}'"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "4Kdnwt8eSIVG"
      },
      "source": [
        "## Look for rare variants with VarScan"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "To check variants risen due to antigenic shift as viral quasispecies, the really low bar should be set (the threshold in 0.1% frequency)"
      ],
      "metadata": {
        "id": "SwfjI-Gb1ADn"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "B4q9YZLTSJ4F",
        "outputId": "77df227e-5ae7-44d5-ae74-13684c3a8f65"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Only SNPs will be reported\n",
            "Warning: No p-value threshold provided, so p-values will not be calculated\n",
            "Min coverage:\t8\n",
            "Min reads2:\t2\n",
            "Min var freq:\t0.001\n",
            "Min avg qual:\t15\n",
            "P-value thresh:\t0.01\n",
            "Reading input from roommate_aligned.mpileup\n",
            "1665 bases in pileup file\n",
            "23 variant positions (21 SNP, 2 indel)\n",
            "0 were failed by the strand-filter\n",
            "21 variant positions reported (21 SNP, 0 indel)\n"
          ]
        }
      ],
      "source": [
        "!varscan mpileup2snp roommate_aligned.mpileup --min-var-freq 0.001 --strand-filter 0 --variants --output-vcf 1 > varscan_results_0001.vcf"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Let's transfer obtained results into files to process in more user-friendly Python"
      ],
      "metadata": {
        "id": "njPbD-bM1e38"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "fSKbIp2DS_-G",
        "outputId": "eda8f82c-884f-490e-e405-a3c2d41ef990"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "KF848938.1 72 A G 1/1:255:16832:16794:6:16787:99.96%:0E0:35:36:4:2:10898:5889\n",
            "KF848938.1 117 C T 1/1:255:20768:20663:36:20625:99.82%:0E0:35:37:27:9:13462:7163\n",
            "KF848938.1 254 A G 0/1:20:35781:35626:35562:59:0.17%:8.5683E-3:36:36:23919:11643:37:22\n",
            "KF848938.1 276 A G 0/1:24:37022:36965:36892:64:0.17%:3.3004E-3:37:35:22579:14313:30:34\n",
            "KF848938.1 307 C T 0/1:255:37506:37386:37029:351:0.94%:6.9068E-66:36:35:22400:14629:184:167\n",
            "KF848938.1 340 T C 0/1:23:37973:37793:37723:64:0.17%:4.6441E-3:37:36:23413:14310:40:24\n",
            "KF848938.1 389 T C 0/1:40:31960:31763:31690:69:0.22%:9.0525E-5:37:36:16067:15623:44:25\n",
            "KF848938.1 691 A G 0/1:23:39503:39319:39245:67:0.17%:4.1867E-3:37:34:21313:17932:23:44\n",
            "KF848938.1 722 A G 0/1:36:38090:38036:37954:76:0.2%:2.353E-4:37:36:21003:16951:39:37\n",
            "KF848938.1 744 A G 0/1:23:38443:38380:38306:65:0.17%:5.0068E-3:37:33:20859:17447:35:30\n",
            "KF848938.1 774 T C 1/1:255:38397:38191:8:38177:99.96%:0E0:32:37:7:1:19806:18371\n",
            "KF848938.1 802 A G 0/1:59:44236:44106:44001:101:0.23%:1.2224E-6:37:35:20575:23426:32:69\n",
            "KF848938.1 859 A G 0/1:25:35895:35721:35655:63:0.18%:3.0291E-3:37:36:14499:21156:27:36\n",
            "KF848938.1 915 T C 0/1:30:37022:36921:36842:69:0.19%:8.2327E-4:35:34:18059:18783:37:32\n",
            "KF848938.1 999 C T 1/1:255:32561:32115:40:32070:99.86%:0E0:35:36:23:17:17184:14886\n",
            "KF848938.1 1043 A G 0/1:25:31638:31590:31529:58:0.18%:2.756E-3:35:33:17353:14176:19:39\n",
            "KF848938.1 1086 A G 0/1:29:25259:25258:25203:53:0.21%:1.0076E-3:36:35:12596:12607:21:32\n",
            "KF848938.1 1213 A G 0/1:34:25177:25093:25035:56:0.22%:3.7244E-4:37:36:8964:16071:24:32\n",
            "KF848938.1 1260 A C 1/1:255:23067:23033:2:23019:99.94%:0E0:32:37:0:2:9824:13195\n",
            "KF848938.1 1280 T C 0/1:20:23487:23462:23418:43:0.18%:9.2875E-3:37:35:11147:12271:24:19\n",
            "KF848938.1 1458 T C 0/1:255:27064:26929:26701:225:0.84%:3.4507E-41:37:35:6865:19836:80:145\n"
          ]
        }
      ],
      "source": [
        "!cat varscan_results_0001.vcf | awk 'NR>24 {print $1, $2, $4, $5, $10}'"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "WjoEBW3P1V6e"
      },
      "outputs": [],
      "source": [
        "!cat varscan_results_0001.vcf | awk 'NR>24 {print $1, $2, $4, $5, $10}' > roommate.txt"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "uZLu2A_v1fdT"
      },
      "outputs": [],
      "source": [
        "with open('roommate.txt') as roommate:\n",
        "  roommate = roommate.readlines()\n",
        "  roommate = ''.join(roommate)\n",
        "\n",
        "b = roommate.replace(' ', ':').split('\\n')\n",
        "res0 = [i.split(':') for i in b][:-1]\n",
        "\n",
        "with open('list_roommate.csv', 'w') as roommate:\n",
        "  for i in res0:\n",
        "    roommate.write(','.join([i[2], i[1], i[3], i[10][:-1]]))\n",
        "    roommate.write('\\n')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "WjnBPMdfYLtX"
      },
      "source": [
        "## Inspect and align the control sample sequencing data"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "RpFmCY-Keq72"
      },
      "source": [
        "### coverage"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Z4ni7V30YSlF"
      },
      "source": [
        "Let's calculate how many reads are in each control, and take a rough estimate of the coverage in your samples\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!gunzip control*"
      ],
      "metadata": {
        "id": "rIeX70OnYdl2"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "df0XJpdGTqBo",
        "outputId": "b9960883-84c2-462e-d453-7266cbc03418"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "  1026344 control1.fastq\n",
            "   933308 control2.fastq\n",
            "   999856 control3.fastq\n",
            "  2959508 total\n"
          ]
        }
      ],
      "source": [
        "! wc -l control*"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "As the length of the reference gene is the sum of all nucleotides"
      ],
      "metadata": {
        "id": "3_ntgXJfiW17"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Cy4lgqIUZ5tx",
        "outputId": "fdd2cb1a-e315-4fff-d447-ce39a6f83e22"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "1665\n"
          ]
        }
      ],
      "source": [
        "!grep -v \">\" reference.fasta | grep -E -o \"G|C|T|A|N\" | wc -l"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "4kE_PI9dYgmD",
        "outputId": "60211ac1-5dcc-4a4c-b01b-970e17d7c0bf"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Number of reads in control #1: 256,586.0\n",
            "Coverage for control #1 is ~154.11\n",
            "Number of reads in control #2: 233,327.0\n",
            "Coverage for control #2 is ~140.14\n",
            "Number of reads in control #3: 249,964.0\n",
            "Coverage for control #3 is ~150.13\n"
          ]
        }
      ],
      "source": [
        "# impliment it via awk?\n",
        "lines = [1026344 / 4, 933308 / 4, 999856 / 4]\n",
        "for i in range(len(lines)):\n",
        "  print(f\"Number of reads in control #{i+1}: {lines[i]:,}\")\n",
        "  print(f'Coverage for control #{i+1} is ~{lines[i] / 1665:.2f}')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "eGTHqehCeuSO"
      },
      "source": [
        "### control"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "bidLbs9CcZVh"
      },
      "source": [
        "To eliminate the noise from sequencing and amplification process, let's implement the same pipeline for control (three random isogenic viral samples)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "oJ78fjZncc7Q"
      },
      "outputs": [],
      "source": [
        "!gunzip control?.fastq.gz"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "PI0ALtRQe0pi",
        "outputId": "604c83db-4e5d-49e5-f22f-4c8a78a253db"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
            "[M::process] read 67448 sequences (10000210 bp)...\n",
            "[M::process] read 67230 sequences (10000094 bp)...\n",
            "[M::mem_process_seqs] Processed 67448 reads in 1.685 CPU sec, 1.624 real sec\n",
            "[M::process] read 67100 sequences (10000113 bp)...\n",
            "[M::mem_process_seqs] Processed 67230 reads in 2.444 CPU sec, 2.371 real sec\n",
            "[M::process] read 54808 sequences (8118313 bp)...\n",
            "[M::mem_process_seqs] Processed 67100 reads in 2.305 CPU sec, 2.365 real sec\n",
            "[M::mem_process_seqs] Processed 54808 reads in 1.966 CPU sec, 2.098 real sec\n",
            "[main] Version: 0.7.17-r1188\n",
            "[main] CMD: bwa mem reference.fasta control1.fastq\n",
            "[main] Real time: 9.371 sec; CPU: 8.522 sec\n",
            "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
            "[M::process] read 67476 sequences (10000274 bp)...\n",
            "[M::process] read 67236 sequences (10000039 bp)...\n",
            "[M::mem_process_seqs] Processed 67476 reads in 1.802 CPU sec, 1.749 real sec\n",
            "[M::process] read 67216 sequences (10000283 bp)...\n",
            "[M::mem_process_seqs] Processed 67236 reads in 2.391 CPU sec, 3.585 real sec\n",
            "[M::process] read 31399 sequences (4635971 bp)...\n",
            "[M::mem_process_seqs] Processed 67216 reads in 2.342 CPU sec, 2.302 real sec\n",
            "[M::mem_process_seqs] Processed 31399 reads in 1.433 CPU sec, 1.536 real sec\n",
            "[main] Version: 0.7.17-r1188\n",
            "[main] CMD: bwa mem reference.fasta control2.fastq\n",
            "[main] Real time: 9.726 sec; CPU: 8.064 sec\n",
            "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
            "[M::process] read 67356 sequences (10000126 bp)...\n",
            "[M::process] read 67208 sequences (10000122 bp)...\n",
            "[M::mem_process_seqs] Processed 67356 reads in 1.675 CPU sec, 1.606 real sec\n",
            "[M::process] read 67010 sequences (10000072 bp)...\n",
            "[M::mem_process_seqs] Processed 67208 reads in 2.440 CPU sec, 2.338 real sec\n",
            "[M::process] read 48390 sequences (7170166 bp)...\n",
            "[M::mem_process_seqs] Processed 67010 reads in 2.252 CPU sec, 2.291 real sec\n",
            "[M::mem_process_seqs] Processed 48390 reads in 1.939 CPU sec, 1.933 real sec\n",
            "[main] Version: 0.7.17-r1188\n",
            "[main] CMD: bwa mem reference.fasta control3.fastq\n",
            "[main] Real time: 8.983 sec; CPU: 8.421 sec\n"
          ]
        }
      ],
      "source": [
        "!bwa mem reference.fasta control1.fastq | samtools view -S -b - | samtools sort -o control1_sorted.bam -\n",
        "!bwa mem reference.fasta control2.fastq | samtools view -S -b - | samtools sort -o control2_sorted.bam -\n",
        "!bwa mem reference.fasta control3.fastq | samtools view -S -b - | samtools sort -o control3_sorted.bam -"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "O8P9vsHZe5Aw",
        "outputId": "f57c1019-e981-49c9-e585-23a7035aedcd"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "256744 + 0 in total (QC-passed reads + QC-failed reads)\n",
            "0 + 0 secondary\n",
            "158 + 0 supplementary\n",
            "0 + 0 duplicates\n",
            "256658 + 0 mapped (99.97% : N/A)\n",
            "0 + 0 paired in sequencing\n",
            "0 + 0 read1\n",
            "0 + 0 read2\n",
            "0 + 0 properly paired (N/A : N/A)\n",
            "0 + 0 with itself and mate mapped\n",
            "0 + 0 singletons (N/A : N/A)\n",
            "0 + 0 with mate mapped to a different chr\n",
            "0 + 0 with mate mapped to a different chr (mapQ>=5)\n",
            "233451 + 0 in total (QC-passed reads + QC-failed reads)\n",
            "0 + 0 secondary\n",
            "124 + 0 supplementary\n",
            "0 + 0 duplicates\n",
            "233375 + 0 mapped (99.97% : N/A)\n",
            "0 + 0 paired in sequencing\n",
            "0 + 0 read1\n",
            "0 + 0 read2\n",
            "0 + 0 properly paired (N/A : N/A)\n",
            "0 + 0 with itself and mate mapped\n",
            "0 + 0 singletons (N/A : N/A)\n",
            "0 + 0 with mate mapped to a different chr\n",
            "0 + 0 with mate mapped to a different chr (mapQ>=5)\n",
            "250184 + 0 in total (QC-passed reads + QC-failed reads)\n",
            "0 + 0 secondary\n",
            "220 + 0 supplementary\n",
            "0 + 0 duplicates\n",
            "250108 + 0 mapped (99.97% : N/A)\n",
            "0 + 0 paired in sequencing\n",
            "0 + 0 read1\n",
            "0 + 0 read2\n",
            "0 + 0 properly paired (N/A : N/A)\n",
            "0 + 0 with itself and mate mapped\n",
            "0 + 0 singletons (N/A : N/A)\n",
            "0 + 0 with mate mapped to a different chr\n",
            "0 + 0 with mate mapped to a different chr (mapQ>=5)\n"
          ]
        }
      ],
      "source": [
        "!samtools flagstat control1_sorted.bam\n",
        "!samtools flagstat control2_sorted.bam\n",
        "!samtools flagstat control3_sorted.bam"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "As we can see the alignment is perfect again"
      ],
      "metadata": {
        "id": "awOQGyi32ohM"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "tlloAF_Ne85t"
      },
      "outputs": [],
      "source": [
        "!samtools index control1_sorted.bam\n",
        "!samtools index control2_sorted.bam\n",
        "!samtools index control3_sorted.bam"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "s6RePzwEcT3Z"
      },
      "source": [
        "## Use VarScan to look for rare variants in the reference files."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "sYifJDWHhG3t",
        "outputId": "0e776d5f-106e-4a93-b2ac-d8c7cd1145c5"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[mpileup] 1 samples in 1 input files\n",
            "[mpileup] Max depth set to maximum value (2147483647)\n",
            "[mpileup] 1 samples in 1 input files\n",
            "[mpileup] Max depth set to maximum value (2147483647)\n",
            "[mpileup] 1 samples in 1 input files\n",
            "[mpileup] Max depth set to maximum value (2147483647)\n"
          ]
        }
      ],
      "source": [
        "!samtools mpileup -f reference.fasta -o control1_aligned.mpileup -d 0 control1_sorted.bam\n",
        "!samtools mpileup -f reference.fasta -o control2_aligned.mpileup -d 0 control2_sorted.bam\n",
        "!samtools mpileup -f reference.fasta -o control3_aligned.mpileup -d 0 control3_sorted.bam"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "voj8GIZkcd-C",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "0f3bccbd-2e37-4339-8e02-19ad637a4641"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Only SNPs will be reported\n",
            "Warning: No p-value threshold provided, so p-values will not be calculated\n",
            "Min coverage:\t8\n",
            "Min reads2:\t2\n",
            "Min var freq:\t0.001\n",
            "Min avg qual:\t15\n",
            "P-value thresh:\t0.01\n",
            "Reading input from control1_aligned.mpileup\n",
            "1665 bases in pileup file\n",
            "58 variant positions (58 SNP, 0 indel)\n",
            "1 were failed by the strand-filter\n",
            "57 variant positions reported (57 SNP, 0 indel)\n",
            "Only SNPs will be reported\n",
            "Warning: No p-value threshold provided, so p-values will not be calculated\n",
            "Min coverage:\t8\n",
            "Min reads2:\t2\n",
            "Min var freq:\t0.001\n",
            "Min avg qual:\t15\n",
            "P-value thresh:\t0.01\n",
            "Reading input from control2_aligned.mpileup\n",
            "1665 bases in pileup file\n",
            "54 variant positions (54 SNP, 0 indel)\n",
            "2 were failed by the strand-filter\n",
            "52 variant positions reported (52 SNP, 0 indel)\n",
            "Only SNPs will be reported\n",
            "Warning: No p-value threshold provided, so p-values will not be calculated\n",
            "Min coverage:\t8\n",
            "Min reads2:\t2\n",
            "Min var freq:\t0.001\n",
            "Min avg qual:\t15\n",
            "P-value thresh:\t0.01\n",
            "Reading input from control3_aligned.mpileup\n",
            "1665 bases in pileup file\n",
            "61 variant positions (61 SNP, 0 indel)\n",
            "0 were failed by the strand-filter\n",
            "61 variant positions reported (61 SNP, 0 indel)\n"
          ]
        }
      ],
      "source": [
        "!varscan mpileup2snp control1_aligned.mpileup --min-var-freq 0.001 --variants --output-vcf 1 > control1_varscan_results_0001.vcf\n",
        "!varscan mpileup2snp control2_aligned.mpileup --min-var-freq 0.001 --variants --output-vcf 1 > control2_varscan_results_0001.vcf\n",
        "!varscan mpileup2snp control3_aligned.mpileup --min-var-freq 0.001 --variants --output-vcf 1 > control3_varscan_results_0001.vcf"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "IkirOd-AyuiE"
      },
      "source": [
        "### table making"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "hLly3ONVnnKt",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "816bc4af-d3f5-482a-c784-b76a0932690b"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "##fileformat=VCFv4.1\n",
            "##source=VarScan2\n",
            "##INFO=<ID=ADP,Number=1,Type=Integer,Description=\"Average per-sample depth of bases with Phred score >= 15\">\n",
            "##INFO=<ID=WT,Number=1,Type=Integer,Description=\"Number of samples called reference (wild-type)\">\n",
            "##INFO=<ID=HET,Number=1,Type=Integer,Description=\"Number of samples called heterozygous-variant\">\n",
            "##INFO=<ID=HOM,Number=1,Type=Integer,Description=\"Number of samples called homozygous-variant\">\n",
            "##INFO=<ID=NC,Number=1,Type=Integer,Description=\"Number of samples not called\">\n",
            "##FILTER=<ID=str10,Description=\"Less than 10% or more than 90% of variant supporting reads on one strand\">\n",
            "##FILTER=<ID=indelError,Description=\"Likely artifact due to indel reads at this position\">\n",
            "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\n",
            "##FORMAT=<ID=GQ,Number=1,Type=Integer,Description=\"Genotype Quality\">\n",
            "##FORMAT=<ID=SDP,Number=1,Type=Integer,Description=\"Raw Read Depth as reported by SAMtools\">\n",
            "##FORMAT=<ID=DP,Number=1,Type=Integer,Description=\"Quality Read Depth of bases with Phred score >= 15\">\n",
            "##FORMAT=<ID=RD,Number=1,Type=Integer,Description=\"Depth of reference-supporting bases (reads1)\">\n",
            "##FORMAT=<ID=AD,Number=1,Type=Integer,Description=\"Depth of variant-supporting bases (reads2)\">\n",
            "##FORMAT=<ID=FREQ,Number=1,Type=String,Description=\"Variant allele frequency\">\n"
          ]
        }
      ],
      "source": [
        "!cat control2_varscan_results_0001.vcf | head -n 16"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "wb7k56chhJG0"
      },
      "outputs": [],
      "source": [
        "!cat control1_varscan_results_0001.vcf | awk 'NR>24 {print $1, $2, $4, $5, $10}' > c1.txt"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "SQo97j3ZsytH"
      },
      "outputs": [],
      "source": [
        "with open('c1.txt') as c1:\n",
        "  c1 = c1.readlines()\n",
        "  c1 = ''.join(c1)\n",
        "\n",
        "b = c1.replace(' ', ':').split('\\n')\n",
        "res1 = [i.split(':') for i in b][:-1]\n",
        "\n",
        "with open('list_control1.csv', 'w') as l1:\n",
        "  for i in res1:\n",
        "    l1.write(','.join([i[2], i[1], i[3], i[10][:-1]]))\n",
        "    l1.write('\\n')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "HgeOFq1WiwnZ"
      },
      "outputs": [],
      "source": [
        "!cat control2_varscan_results_0001.vcf | awk 'NR>24 {print $1, $2, $4, $5, $10}' > c2.txt"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "_-D8U3FnuTEv"
      },
      "outputs": [],
      "source": [
        "with open('c2.txt') as c2:\n",
        "  c2 = c2.readlines()\n",
        "  c2 = ''.join(c2)\n",
        "\n",
        "b = c2.replace(' ', ':').split('\\n')\n",
        "res2 = [i.split(':') for i in b][:-1]\n",
        "\n",
        "with open('list_control2.csv', 'w') as l2:\n",
        "  for i in res2:\n",
        "    l2.write(','.join([i[2], i[1], i[3], i[10][:-1]]))\n",
        "    l2.write('\\n')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "emA5Pte_iyTX"
      },
      "outputs": [],
      "source": [
        "!cat control3_varscan_results_0001.vcf | awk 'NR>24 {print $1, $2, $4, $5, $10}' > c3.txt"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "y0CniBiWvxSX"
      },
      "outputs": [],
      "source": [
        "with open('c3.txt') as c3:\n",
        "  c3 = c3.readlines()\n",
        "  c3 = ''.join(c3)\n",
        "\n",
        "b = c3.replace(' ', ':').split('\\n')\n",
        "res3 = [i.split(':') for i in b][:-1]\n",
        "\n",
        "with open('list_control3.csv', 'w') as l3:\n",
        "  for i in res3:\n",
        "    l3.write(','.join([i[2], i[1], i[3], i[10][:-1]]))\n",
        "    l3.write('\\n')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "w5kjOJA1zr34"
      },
      "source": [
        "## Compare the control results to your roommate’s results"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Finally, let's visualise our results for test and control groups"
      ],
      "metadata": {
        "id": "aYque_bk2whx"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "anyo0lL32Njq"
      },
      "outputs": [],
      "source": [
        "import numpy as np\n",
        "import pandas as pd\n",
        "import matplotlib.pyplot as plt"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "pfoR8dukzvjD"
      },
      "outputs": [],
      "source": [
        "test = pd.read_csv('list_roommate.csv', names=['org', 'pos', 'alt', 'cov'])\n",
        "con1 = pd.read_csv('list_control1.csv', names=['org', 'pos', 'alt', 'cov'])\n",
        "con2 = pd.read_csv('list_control2.csv', names=['org', 'pos', 'alt', 'cov'])\n",
        "con3 = pd.read_csv('list_control3.csv', names=['org', 'pos', 'alt', 'cov'])"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "To establish noise boundaries, the confidence intervals of controls should be calculated as $\\mu \\pm 3 \\sigma$ for 99.7% probability "
      ],
      "metadata": {
        "id": "yrUO8QGb23cP"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "cons = [con1, con2, con3]\n",
        "for i in range(3):\n",
        "  sd, meany = cons[i][\"cov\"].std(), cons[i][\"cov\"].mean()\n",
        "  print(f'For control #{i+1} the sd is {sd:.3f}% and the average is {meany:.3f}%')\n",
        "  print(f'So, the 3x sd deviation is ({meany-3*sd:.3f}, {meany+3*sd:.3f})%')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "DVUvUU6DoX4j",
        "outputId": "3e9d8099-32dd-4b17-c619-47a092221d47"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "For control #1 the sd is 0.072% and the average is 0.256%\n",
            "So, the 3x sd deviation is (0.041, 0.472)%\n",
            "For control #2 the sd is 0.052% and the average is 0.237%\n",
            "So, the 3x sd deviation is (0.080, 0.394)%\n",
            "For control #3 the sd is 0.078% and the average is 0.250%\n",
            "So, the 3x sd deviation is (0.016, 0.484)%\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "comb_cons = list(con1[\"cov\"]) + list(con2[\"cov\"]) + list(con3[\"cov\"])\n",
        "lower = np.mean(comb_cons) - 3* np.std(comb_cons)\n",
        "upper = np.mean(comb_cons) + 3* np.std(comb_cons)\n",
        "print(f'The combined control has 99,7% CI as ({lower:.3f}, {upper:.3f})%')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "3FTpxav5wxPe",
        "outputId": "8e805015-ff87-421b-c992-7e8859105bbb"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "The combined control has 99,7% CI as (0.042, 0.455)%\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Let's see processed test results and eliminate ones with noise-level signal"
      ],
      "metadata": {
        "id": "kVKxW7mygmmb"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "test.transpose()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 198
        },
        "id": "srH7ZZ4UJkvz",
        "outputId": "ef24e505-4057-4f81-ea72-906a5d4a0c4f"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "        0      1     2     3     4     5     6     7    8     9   ...    11  \\\n",
              "org      A      C     A     A     C     T     T     A    A     A  ...     A   \n",
              "pos     72    117   254   276   307   340   389   691  722   744  ...   802   \n",
              "alt      G      T     G     G     T     C     C     G    G     G  ...     G   \n",
              "cov  99.96  99.82  0.17  0.17  0.94  0.17  0.22  0.17  0.2  0.17  ...  0.23   \n",
              "\n",
              "       12    13     14    15    16    17     18    19    20  \n",
              "org     A     T      C     A     A     A      A     T     T  \n",
              "pos   859   915    999  1043  1086  1213   1260  1280  1458  \n",
              "alt     G     C      T     G     G     G      C     C     C  \n",
              "cov  0.18  0.19  99.86  0.18  0.21  0.22  99.94  0.18  0.84  \n",
              "\n",
              "[4 rows x 21 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-e63cda96-2d29-4184-a8ea-fd3a780722a2\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>0</th>\n",
              "      <th>1</th>\n",
              "      <th>2</th>\n",
              "      <th>3</th>\n",
              "      <th>4</th>\n",
              "      <th>5</th>\n",
              "      <th>6</th>\n",
              "      <th>7</th>\n",
              "      <th>8</th>\n",
              "      <th>9</th>\n",
              "      <th>...</th>\n",
              "      <th>11</th>\n",
              "      <th>12</th>\n",
              "      <th>13</th>\n",
              "      <th>14</th>\n",
              "      <th>15</th>\n",
              "      <th>16</th>\n",
              "      <th>17</th>\n",
              "      <th>18</th>\n",
              "      <th>19</th>\n",
              "      <th>20</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>org</th>\n",
              "      <td>A</td>\n",
              "      <td>C</td>\n",
              "      <td>A</td>\n",
              "      <td>A</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>T</td>\n",
              "      <td>A</td>\n",
              "      <td>A</td>\n",
              "      <td>A</td>\n",
              "      <td>...</td>\n",
              "      <td>A</td>\n",
              "      <td>A</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>A</td>\n",
              "      <td>A</td>\n",
              "      <td>A</td>\n",
              "      <td>A</td>\n",
              "      <td>T</td>\n",
              "      <td>T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>pos</th>\n",
              "      <td>72</td>\n",
              "      <td>117</td>\n",
              "      <td>254</td>\n",
              "      <td>276</td>\n",
              "      <td>307</td>\n",
              "      <td>340</td>\n",
              "      <td>389</td>\n",
              "      <td>691</td>\n",
              "      <td>722</td>\n",
              "      <td>744</td>\n",
              "      <td>...</td>\n",
              "      <td>802</td>\n",
              "      <td>859</td>\n",
              "      <td>915</td>\n",
              "      <td>999</td>\n",
              "      <td>1043</td>\n",
              "      <td>1086</td>\n",
              "      <td>1213</td>\n",
              "      <td>1260</td>\n",
              "      <td>1280</td>\n",
              "      <td>1458</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>alt</th>\n",
              "      <td>G</td>\n",
              "      <td>T</td>\n",
              "      <td>G</td>\n",
              "      <td>G</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>C</td>\n",
              "      <td>G</td>\n",
              "      <td>G</td>\n",
              "      <td>G</td>\n",
              "      <td>...</td>\n",
              "      <td>G</td>\n",
              "      <td>G</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>G</td>\n",
              "      <td>G</td>\n",
              "      <td>G</td>\n",
              "      <td>C</td>\n",
              "      <td>C</td>\n",
              "      <td>C</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>cov</th>\n",
              "      <td>99.96</td>\n",
              "      <td>99.82</td>\n",
              "      <td>0.17</td>\n",
              "      <td>0.17</td>\n",
              "      <td>0.94</td>\n",
              "      <td>0.17</td>\n",
              "      <td>0.22</td>\n",
              "      <td>0.17</td>\n",
              "      <td>0.2</td>\n",
              "      <td>0.17</td>\n",
              "      <td>...</td>\n",
              "      <td>0.23</td>\n",
              "      <td>0.18</td>\n",
              "      <td>0.19</td>\n",
              "      <td>99.86</td>\n",
              "      <td>0.18</td>\n",
              "      <td>0.21</td>\n",
              "      <td>0.22</td>\n",
              "      <td>99.94</td>\n",
              "      <td>0.18</td>\n",
              "      <td>0.84</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>4 rows × 21 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-e63cda96-2d29-4184-a8ea-fd3a780722a2')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-e63cda96-2d29-4184-a8ea-fd3a780722a2 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-e63cda96-2d29-4184-a8ea-fd3a780722a2');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 55
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Above the established level, highly likely to be true data is:"
      ],
      "metadata": {
        "id": "wNA8qsqh3ptQ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "test[test['cov']>0.484].transpose()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 170
        },
        "id": "U3Ha0dyggODP",
        "outputId": "fa69d03e-03f8-437a-a7f3-619ff245b393"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "        0      1     4      10     14     18    20\n",
              "org      A      C     C      T      C      A     T\n",
              "pos     72    117   307    774    999   1260  1458\n",
              "alt      G      T     T      C      T      C     C\n",
              "cov  99.96  99.82  0.94  99.96  99.86  99.94  0.84"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-1657f2a7-224c-4f1b-9bd9-30f2b0c49c4d\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>0</th>\n",
              "      <th>1</th>\n",
              "      <th>4</th>\n",
              "      <th>10</th>\n",
              "      <th>14</th>\n",
              "      <th>18</th>\n",
              "      <th>20</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>org</th>\n",
              "      <td>A</td>\n",
              "      <td>C</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>A</td>\n",
              "      <td>T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>pos</th>\n",
              "      <td>72</td>\n",
              "      <td>117</td>\n",
              "      <td>307</td>\n",
              "      <td>774</td>\n",
              "      <td>999</td>\n",
              "      <td>1260</td>\n",
              "      <td>1458</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>alt</th>\n",
              "      <td>G</td>\n",
              "      <td>T</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>C</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>cov</th>\n",
              "      <td>99.96</td>\n",
              "      <td>99.82</td>\n",
              "      <td>0.94</td>\n",
              "      <td>99.96</td>\n",
              "      <td>99.86</td>\n",
              "      <td>99.94</td>\n",
              "      <td>0.84</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-1657f2a7-224c-4f1b-9bd9-30f2b0c49c4d')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-1657f2a7-224c-4f1b-9bd9-30f2b0c49c4d button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-1657f2a7-224c-4f1b-9bd9-30f2b0c49c4d');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 56
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Probably (99,7%) noise in the test is:"
      ],
      "metadata": {
        "id": "bregigjX3vou"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "test[test['cov']<=0.484].transpose()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 170
        },
        "id": "fuxU75lYgdlD",
        "outputId": "2c452a40-2864-4858-d508-c5f51612ae2e"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "       2     3     5     6     7    8     9     11    12    13    15    16  \\\n",
              "org     A     A     T     T     A    A     A     A     A     T     A     A   \n",
              "pos   254   276   340   389   691  722   744   802   859   915  1043  1086   \n",
              "alt     G     G     C     C     G    G     G     G     G     C     G     G   \n",
              "cov  0.17  0.17  0.17  0.22  0.17  0.2  0.17  0.23  0.18  0.19  0.18  0.21   \n",
              "\n",
              "       17    19  \n",
              "org     A     T  \n",
              "pos  1213  1280  \n",
              "alt     G     C  \n",
              "cov  0.22  0.18  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-3ac319ae-a01e-4139-8d90-ac6e68b1ab81\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>2</th>\n",
              "      <th>3</th>\n",
              "      <th>5</th>\n",
              "      <th>6</th>\n",
              "      <th>7</th>\n",
              "      <th>8</th>\n",
              "      <th>9</th>\n",
              "      <th>11</th>\n",
              "      <th>12</th>\n",
              "      <th>13</th>\n",
              "      <th>15</th>\n",
              "      <th>16</th>\n",
              "      <th>17</th>\n",
              "      <th>19</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>org</th>\n",
              "      <td>A</td>\n",
              "      <td>A</td>\n",
              "      <td>T</td>\n",
              "      <td>T</td>\n",
              "      <td>A</td>\n",
              "      <td>A</td>\n",
              "      <td>A</td>\n",
              "      <td>A</td>\n",
              "      <td>A</td>\n",
              "      <td>T</td>\n",
              "      <td>A</td>\n",
              "      <td>A</td>\n",
              "      <td>A</td>\n",
              "      <td>T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>pos</th>\n",
              "      <td>254</td>\n",
              "      <td>276</td>\n",
              "      <td>340</td>\n",
              "      <td>389</td>\n",
              "      <td>691</td>\n",
              "      <td>722</td>\n",
              "      <td>744</td>\n",
              "      <td>802</td>\n",
              "      <td>859</td>\n",
              "      <td>915</td>\n",
              "      <td>1043</td>\n",
              "      <td>1086</td>\n",
              "      <td>1213</td>\n",
              "      <td>1280</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>alt</th>\n",
              "      <td>G</td>\n",
              "      <td>G</td>\n",
              "      <td>C</td>\n",
              "      <td>C</td>\n",
              "      <td>G</td>\n",
              "      <td>G</td>\n",
              "      <td>G</td>\n",
              "      <td>G</td>\n",
              "      <td>G</td>\n",
              "      <td>C</td>\n",
              "      <td>G</td>\n",
              "      <td>G</td>\n",
              "      <td>G</td>\n",
              "      <td>C</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>cov</th>\n",
              "      <td>0.17</td>\n",
              "      <td>0.17</td>\n",
              "      <td>0.17</td>\n",
              "      <td>0.22</td>\n",
              "      <td>0.17</td>\n",
              "      <td>0.2</td>\n",
              "      <td>0.17</td>\n",
              "      <td>0.23</td>\n",
              "      <td>0.18</td>\n",
              "      <td>0.19</td>\n",
              "      <td>0.18</td>\n",
              "      <td>0.21</td>\n",
              "      <td>0.22</td>\n",
              "      <td>0.18</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-3ac319ae-a01e-4139-8d90-ac6e68b1ab81')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-3ac319ae-a01e-4139-8d90-ac6e68b1ab81 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-3ac319ae-a01e-4139-8d90-ac6e68b1ab81');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 57
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "M3G02psT2YE1",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 264
        },
        "outputId": "d82c02b4-36cb-4904-be67-9b7d29ec1918"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "Text(0.5, 1.0, 'Mutation frequency in our roommate and controls')"
            ]
          },
          "metadata": {},
          "execution_count": 58
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<Figure size 1296x216 with 1 Axes>"
            ],
            "image/png": "iVBORw0KGgoAAAANSUhEUgAABCkAAADoCAYAAADCDb5lAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAAgAElEQVR4nOzdeXxU1f3/8dcnYU9CAiQBxEBYC7iAGheoVpZSF2pFwCrKYktFVPpDRapflYoKtPTbql1sES1lcSniF5G6VAUBtV8Qg7h8FZEtLEJIWAKBJCzJ+f1x78RhmCQTmCzg+/l4zGMy9557z2funLlwP3POueacQ0RERERERESkpsXUdAAiIiIiIiIiIqAkhYiIiIiIiIjUEkpSiIiIiIiIiEitoCSFiIiIiIiIiNQKSlKIiIiIiIiISK2gJIWIiIiIiIiI1ApKUoiInMLMbKKZOTNbV8b6df76iZXcb6q/7/QTiKmev233kOXpfiw/ruw+T5SZfd/MPjazIjPTPbcrycyyzOz3NR2HVJ6Z/dTMbqnpOCJlZmNq03fUzC6q7Hkzgn3e4p8D46O5XxGR042SFCIip74ioK2ZZQQvNLMLgXR/fWWlAg/721dWPX/b7iHLdwA9gA9OYJ8n6mkgD7jCr1sq5zrgTzUdhJyQnwK31HQQp7CL8M5jIiJSzZSkEBE59R0E3gVuDFl+o7/8YLVHFIZz7pBzboVzLq8aq+0MLHDOLXPOrQhXwMwaVmM8pxTn3Grn3JaajqM85mkQ7bIikVCbEhGJPiUpREROD/8EfmpmBt5/nPF+Sf1naEEzW2pmL4cs6+V3Qz7bH+Lxub9qib/c+eXizOwvZrbWzArMbJOZPWVmjYN2l+8//yOwrT/U47jhHmYW6w8N2WJmh8zsCzO7KSS2mWaWaWb9zOwzMztoZh+Y2VllHYzA+wFigT/69c7012WZ2R/MbIKZbQP2+8tjzOx+M1vvx/K1mY0I2a/58eaYWb6ZzTazmwLvMfRYRnDcLzOzZf6x3G1mz5hZQtD6QPfwc8zsHf+9f2VmA8O85+vMbKWZFfr7esPM2phZV38fvULKx5vZATMbW85xPGa4x4l8FkHbtjWzBWa23z92/zKzDkHrww4HCtQZ9Hqime0ys0vN7CO8nkLXl1FnmWXNGw7xuf9ZbzWzyWZWJ2T77ma22P989prZ82bWPEzMN5rZP/z3ts3Mhvrrf2Vm280s18ymmllMmNgu9o9poX8s25o33GqB//msMbM+IXEN98vu8eNaYkE9qfy2Pgi43L79Dk4MWn+tX2eRmWWb2e/MrG4Fn19/vw3m+O9zhZn9qIzjfZ6/vsDMVpvZZSHl6pt3Hsnz38MTQLn1B217rt928vzjs9LM+gWtL7ed+WWcmY01syn+Z5Nj3nmsvr/+FuDPQWWdmS0NeY8n1KbCvJ//Mu+cU2RmO83s32bWIpJjISJyulKSQkTk9DAfaA5c6r++DEjxl1fWDuBm/+878YZJBIZKNMK78H8QuAqYAPQB5gVtH7igmhS07Y4y6nrU39d04CfAf4DnzWxISLnWwH8Dk4EheMNR5pp5SZkwPg6K+Q/+348Frb8JuBy4A7jBX/Zn4CE/lv7AK8CMkIvm/wf82i8zGCgEfldGDOUys+8Di4Bsf193AVcD/whT/AVgId7wi3XAP83szKB9DcP7rDfgJad+BnwNpDjnvgRWcHzX/+vxLgyfq2Tolf0s8C/+FgNdgFv9WNoCy8ysaSXrB68dzgKeBa4EVlamrH9xPRevnVyL99nfC/wlKOYUYKm//U3AL/HazDtmVi+kjql4bXwQ8D4wy8z+gDdk4OfAk8Cv8D6b0NimA0/gHcvWwBzgRbxhUQOBb4B5ZtYoaLt0YDbeZ3gTsBV438za+esfA5YAq/n2O/is/75+itdWVuJ95x4BRgG/KecYgvd5/QsY5r/P/wXe9Ntx6HuahTfUahBwCJgfEv9vgV/4cd4MtAHGVVA/ZtYZ7xzREhiN9314BUjz11emnY0DzgCG4rXn24BAwu51vPMGfHv87gjzHivVpsK8n+HAA8DjeEPSbgfWA3EVHQsRkdOac04PPfTQQ49T9AFMBHb5f78KPOX//Ve8YQ4Au4CJQdssBV4O2U8vwAFn+6/P9l/3qqD+OsD3/bKt/WXx/utbQsqm+8t/7L9uijcU5eGQcm8Aa4NezwSOAh2Dlg3w99W5gvgcMCZkWRbeBWWDoGUdgBJgREjZ2cBH/t+xwHbgbyFl3vHrSQ93LMs67ngXs0tCyvQJ+Rxu8V//PKhMM/94jPZfx+BdyM4v5zj8AjgAxActey+0HYTZLgv4/cl+FngXlEeBdkHLzgQOA/8Vrn2E1JkZ0uYdcG2E34/jyuIlbUKP/a+AYuBM//Vv8eYzaRxU5mJ/f0NCYv5HUJnGwBG8ZFJs0PKVwNwwsV0etOwOf9mvg5Z19ZddVcZ7jMH7Hn4Vst3LwNKQsgZsDo7XX/5zvIRbs4qOaUidbwEzwrynPkHLuvvLrgxqv4XAfSH7+wpwFdT7IrANaHii7cxf5oD3QrZdAKwIej0mXDwn2aZu8beN91//BfifSI65Hnroocd36aGeFCIip49/AoP9XxMHE2aoRzSY2TC/C/cBvIuxwESYnSq5q7PxfpGcF7J8LtDJ/yU7IMs5F3wHky/95zM5MYudc8ETivbFS1K8YmZ1Ag+8X2W7m1ks3q+1LfGSQcEq3VvF/1W5B/BSSH0f4B3TC0I2eTvwh3NuN5DDt+/9e3i/CIfrgREw138OdEtvj9frprxtynIin8VFwMfOuY2BBc65bXi/il9a5lZlc8CbJ1LW/yzPJ3y7i+HbHjgXAW875/YHxfwhXuImNObFQWX2A7nAMudccVCZ9UCrkO0O4yWrgsuAN5dM6LLSbc2si5m9YmY78S6Cj+C1g4q+g53wemuEtrt3gQZ438mwzOxMM5tlZt/gJQKOAD8KU+dhvIRcQGj7OMevq/R75Jwr4fjvVTh98BI9hWWsr0w7ezvk9ZdEfj450TYV6hPgajN7xLy7icRGWL+IyGlNSQoRkdPHQrxeDJPxugv/K9oVmNl1eL0LluNd8F6C1+UavAuPymjpP+8MWR54Hdw9O3SyzcMnWGdoHQHJeD0l9uFdfAUeM/F+MW4JBMaJ54RsG/o6Ek38+v4aUt8hvCEYaSHlw73/wHtv5j+XNaQG51w+8BLeMBDwftHNBv59ArGfyGfRkuOPOf6yExnusdc5d7jiYmHLJuMd44raXWViDndMyvvMAvL9C/TgMsfsLyj2BgDmzVnyNl4buQdvaNeFwKdh9h8q2X9+g2Pb3SZ/eWi7w68zBu/80hNvuFNvv843K3pPofFzct+jZpTTzjn5zyzS88mJtqlQM/CGe/wU+BDYaWaTlKwQke+6cifzERGRU4dz7qCZvQbcDcxzzpV1V48ivNuEBmsSYTXXAx8650rHZ5vZ5ZUO1hO42EgFdgctD0xMuOcE9xsJF/J6D96vw9/H61ERKodv/81MDVkX+jrQQyPcMd7l/53nxzAR74Ix1PZwQZchcOxallvKGz//gZl1BIYDs0N+6a9KO4Bwk2s259vPubzjFir08ytPaNldeBfmoZ9baLvbEaZMoNyqStQfbT3wfvHv55z7KrDQzBIj2Dbw3kbhzVcRalOYZeANhzoPb8hJaWLLTuzOONn+cyrHfsfDHetQuym/nUfSzqLhRNvUsTvxkjlPAE+YWRre/ByT8Ya0TItatCIipxj1pBAROb38Da8HRXn/wd2Gd2vOYD8KeV3Wr+MN8X7tD3ZzyOtIezn8H1DA8Xdm+CnwtXMut4Lto+ldvJ4Nic65zDCPw3iTE2bjTYoXLPROG9v85y6BBf4FSOkx9xNIK4DvlVFfZZIUa/HmpBhRXiHn3P/6ZWfgdfmfWYk6TtaHwAVm1jawwMxa4f0yHxgulIN3oRd83OL9MlHjJ2ZWEb7dleD1EgrEfIUde7eVC/HmofiAmhNIDJR+D82sJ15cwcL1DAi0lfQy2t1uwgtXZxu8pF5lfY6XkCr9Hvk9NUK/V+EsxruLUVnnlkjaWaQO+9tX2LuiEm2qvH1sdc79Fm94T9dKxioiclpRTwoRkdOIc24px44HD+cVYKR/27/X8bpuXxlSZgve5HYjzGwfcMQ5l4k3SeRTZvYg3gXB1XjzOQTHcNjMNuFdTPwf3gXJZ2Fi3WNmTwIPmdlRIBPvgv9qvDsdVBvn3Fozm4Z314zf+bE0wPtVtpNz7hfOuWJ/3e/NbBfeXAKDCLqo9ve1zbxbZj5mZgV4Pwg8wPG/pv4KWGxmJXiTHObjJQ/6Aw86576OMPYSM/sV3l1RnsebXNDhjd9/0f/cAv6OdyeD5cG/wleDmcB9eHeD+DXePAoP4/0C/TSUvo9XgbvNbDNeb5NxeO0w2h4G3jKzf+DN3XIO3p0mnvHnMADvjgu3++Wm4g2l+i3eRfb/VEFMkVqBNwnqM357PBOvR843IeW+Aq41swF4ibPtzrntZjYOmGPebYPfxLsYb4c3Aepg51xBmDq/8vfxBzObACTg3RUktM4KOed2m9l04BH/e/8F3p044iPY/BHgI+A9/+4pu/F6eOx2zs0ggnZWCYHvx1gzexfY75xbW075SNrUMczsabzzwgq8oWa9gY7+exAR+c5STwoRke8Y59zreBfNg/ESFm349tZ7gTJFeBcOFwDL8C4MwPuP/h/88vP9bW8KU81ovHHai/xtzygjnF/j3frwduA14AfAUOdclUz6WYE78S4qhuMNwZiJlzB4L6jMk8AUvPf3P3gXVr8Ks68heIme5/zyj+L9il3KOfcB3vtNwbvt5L/8fW0l/Lj6MjnnXsBLmHTGS3jM9v8O7Y2ywH+eUZn9nyzn3CHgh3gXfn/Hu33jFry7xwQnb8bgTXL4V+ApvITLu0SZc+5t4EYgA++434XXrscElcnFu2gs8uN4Ci8x1a8S82FEnXNuJ94v9i3wJpu8C689rg8p+le8uStm4H0HR/nbz8XrtdAdb6LH+Xh3FfmYb3tBhdZ5CC+BeBSvfT2G971ddoJv41d+XL/GO7bb8ZJC5fKTBJfiJR2exTt/Dca7Y0ll2lkk3sdL6I3FS8iWm+SIpE2FsRzvHPAPvHPOdcCtzrkF5WwjInLaM+cqM6xTREREgpnZj/EuSto657JqOJxymdkdwO+AM4LvWiEiIiJSW2i4h4iIyGnOzNLxbhX5ADBTCQoRERGprTTcQ0RE5PQ3EW84zRpgQs2GIiIiIlI2DfcQERERERERkVpBPSlEREREREREpFZQkkJEREREREREaoXTduLM5ORkl56eXtNhiIiIiIiIiEiQVatW7XLOpYRbd9omKdLT08nMzKzpMGpOcT7k/xMSboTYhNOnLhERqR107pcAtQWpSmpfIqclM9tc1joN9zgdFefDzjthz5Pec3H+6VGXiIjUDjr3S4DaglQltS+R7yQlKU43gZP54TVQp633XFUn9eqsS0REaged+yVAbUGqktqXyHeWkhSnk+CTeUxzMPOeq+KkXp11iYhI7aBzvwSoLUhVUvsS+U5TkuJ0Ee5kDlVzUq/OukREpHbQuV8C1BakKql9iXznmXOueioyGwPcApwDvOicuyVoXV/gKaA18CFwi3Nus7+uPvA3YDBQAPzOOfd4RfVlZGS478zEmWWdzIM5ByU7oV4XaP7UiU88VJ11iYhI7aBzvwSoLUhVUvsS3/79+8nJyeHIkSM1HYqcoLp165Kamkrjxo3DrjezVc65jLDrqjFJMRAoAa4AGgaSFGaWDGwAfgH8C3gMuMw5d4m//jfApcBPgBbAErwkxr/Lq+87k6SI5GQecLIn9eqsS0REaged+yVAbUGqktqX+Pbv38/OnTtp1aoVDRs2xMprC1IrOecoLCzkm2++oXnz5mETFeUlKaptuIdzbr5zbgGwO2TVQOAL59w851wRMBHoZmad/fUjgMecc3udc2uAZ/B6ZEhlTuZwct3kqrMuERGpHXTulwC1BalKal8SJCcnh1atWtGoUSMlKE5RZkajRo1o1aoVOTk5ld6+NsxJcRbwaeCFc+4gXs+Ks8ysCdAyeL3/91nVGmFtVNmTecCJnNSrsy4REakddO6XALUFqUpqXxLiyJEjNGzY8OR2UpwPec+oXdSwhg0bntCQndqQpIgH9oUs2wck+OsIWR9YdxwzG2VmmWaWmZubG/VAa5X8f0LRqsqdzAMCJ/WiVd5+alNdIiJSO+jcLwFqC1KV1L4kjJPqQRFIfO15UgmsGnain2NtSFIcAEIHqTQG8v11hKwPrDuOc266cy7DOZeRkpIS9UBrlYQbocEF3ni8ys4rEhjH1+ACbz+1qS4REakddO6XALUFqUpqXxJNwT1z6rRVT5tTVG1IUnwBdAu8MLM4oD3ePBV7gR3B6/2/v6jWCGuj2ARvoqB6XSp3Uj+RiYaqsy4REakddO6XALUFqUpqXxIt4YYOaUjQKanakhRmVsfMGgCxQKyZNTCzOsArwNlmNshf/2vgM+fcV/6ms4GHzKyJP5nmrcDM6oq7VqvsSf1kTubVWZeIiNQOOvdLgNqCVCW1LzlZZc1tokRFpS1dupQ6derUaAzV2ZPiIaAQuB8Y6v/9kHMuFxgETAb2AhcDwf21HsabSHMzsAz474puP/qdEulJPRon8+qsS0REaged+yVAbUGqktqXnKiKJl+txkTFiUwSKcerzluQTnTOWchjor9ukXOus3OuoXOul3MuK2i7Q865nzvnGjvnmjvnHq+umE8ZFZ3Uo3kyr866RESkdtC5XwLUFqQqqX1JZUV6d5gqSlSkp6fz6KOP0rt3b+Lj45k3bx5jx44lLS2N5ORkBgwYwJYtW0rLFxQUlLu+V69e3HPPPVx33XUkJCTQvn17Fi9ezKJFizj77LNp3Lgx1113Hfn538ZvZvzlL38hIyODuLg4evbsybZt23jiiSdIS0ujWbNmPPjgg8fEMHDgQFq0aEHjxo05//zzeeeddwDYvn07V111FcXFxcTHxxMfH8+sWbMA2LJlC4MHD6ZFixa0bNmSUaNGHRNHNNWGOSkkGso6qVfFybw66xIRkdpB534JUFuQqqT2JZGq7O1rqyhR8cwzz/D444+Tn5/PsmXLWLFiBStWrGDz5s0kJydzzTXXUFxcDMDdd99d7nqAOXPmcP/995OXl8cNN9zAsGHDmD59Ou+99x5ZWVmsXbuWP/3pT8fE8Nxzz7FgwQJyc3Np0KABffr0Ye/evWzYsIF3332X3//+9/znP/8BoKSkhIEDB7Ju3Tp2797NkCFDGDRoELm5uZxxxhm8+eabxMbGcuDAAQ4cOMCIESMoKiqiT58+dO3alU2bNvHll1+ybds2xo4dG5VjGEpJitNJuJN6VZ3Mq7MuERGpHXTulwC1BalKal9SkcomKAKqIFFx6623ct555+GcY9asWUyaNIlWrVoRFxfHk08+yZo1a1i5ciUlJSXlrg/46U9/ysUXX0xsbCxDhw5lx44djB8/nqZNm9K0aVN+/OMfk5mZeUwM48aN48wzz6RRo0YMHjyY7OxsJk6cSL169ejWrRvdunUr3SY+Pp6hQ4eSkJBA3bp1GT9+PPXq1eOjjz4q8z2+9tprOOd49NFHadiwIU2aNOGxxx7j+eefPybBEi2VSlKYWaqZPWdm/2dmC8ysU9QjkpMTfFI/uqlqT+bVWZeIiNQOOvdLgNqCVCW1LylP/j+haFXlEhQBgURF0SpvPycpPT0dgNzcXA4dOkTbtm1L18XHx5OamsrWrVsrXB/QsmXL0r8bNWoUdlnoMIvQ9ampqcTExITdprCwkDFjxtCuXTsaN25MUlISe/fuJTc3t8z3uGnTJrZs2UJSUlLpo2/fvpgZ2dnZER2nyqjstJ1/Ad4BHgP6Ay8AGdEOSk5S4KSe/0/vntFVeTKvzrpERKR20LlfAtQWpCqpfUlZEm6Egvcr35MCvu2Z0+ACbz8nKZAMSElJoX79+mRlZdGhQwcADhw4QE5ODmlpaRWury6PP/447733HosXLyY9PR0zIzk5GecPrQpObgS0adOGTp068cUXX1RLjOX2pDCzx/3bggakAs8659YCfwXaV2VwchJiEyDp1uo5mVdnXSIiUjvo3C8BagtSldS+JJzK3rY2oAqHDsXExDB8+HAmTJjA9u3bKSgoYNy4cXTu3JmLLrqowvXVZf/+/dSvX59mzZpx+PBhHn30UfLy8krXt2jRguLiYjZt2lS67Mc//jGHDx9mypQp5Ofn45zjm2++4ZVXXqmSGCsa7pEDfGRml/uvFwPvmNlk/++qiUpERERERESkLJVNVFTD3CZPPPEEGRkZXHjhhbRu3ZodO3awcOFCYmNjI1pfHe655x6SkpI444wzaN++PY0aNSodsgLQqVMnbr/9di666CKSkpKYM2cOjRo14t133+XLL7+kc+fOJCYm0rdvXz755JMqidFcBR+mmX0PeAb4EhgP9AC6ARuABc65kiqJ7CRlZGS40AlFREREREREpPZas2YNXbp0iXyDSCbR1OSrNaasz9PMVjnnwk4dUeHEmc65tc65HwBfAB8Csc65/3bOza+tCQoRERERERH5DqioR4USFKecCpMUZtbdzAYDbwNXAXeZ2Rwza1Ll0YmIiIiIiIiUp6xEhRIUp6SKJs58FJgPDAJeB37knLsCWAL8r5+8EBEREREREak54RIVSlCckirqSXE7cL5zbghwMTAawDk3A+gD3Fy14YmIiIiIiIhEIDhRcXSTEhSnqIqSFHuAnmZWF7gU2B1Y4Zzb4Zy7riqDExEREREREYlYIFHR9C4lKE5RdSpY/3Pgr8A84FNgZJVHJCIiIiIiInKiYhMg6daajkJOULlJCufcf/BuNyoiIiIiIiIiUqUqvLuHiIiIiIiIiEh1UJJCRERERERE5DRlZnzwwQc1HUbElKQQERERERERqUVmzpxJhw4dqr3e7t27k5mZCUBKSgrZ2dml6woLC7n++uvp2LEjMTExTJo0qUpiUJJCRERERERE5BRz5MiRqO7v4MGDbNiwgW7durFu3Tri4uJo0aJF6Xozo2fPnkyfPp2LLrooqnUHU5JCREREREREpJIOHDjAvffeS7t27UhISKBr1668//77ABQUFDB27FjS0tJITk5mwIABbNmypXTbXr16MW7cOAYNGkRCQgLt27fn1VdfBWD58uWMHj2ajRs3Eh8fT3x8PEuXLmXp0qXUqVOHOXPm0K5dO5o2bQrAZ599Rp8+fWjSpAnt2rVj0qRJFBcXV/r9ZGZmcu6551K3bl2WL1/OxRdffMz6Bg0acPfdd9O7d28aNGhwooetQiecpDCzTWb2dzNrdbJBmFm6mb1hZnvNLNvM/mJmdfx13c1slZkV+M/dT7Y+EREREREROX2sXr2aRYsWhV23aNEiVq9eHfU6R44cyYcffsjixYvZv38/CxcupGXLlgDcfffdrFixghUrVrB582aSk5O55pprjkkezJo1i3HjxrFv3z7GjBnDiBEjKCgooEePHkybNo127dpx4MABDhw4QK9evQAoLi7mjTfeYPXq1ezcuZN9+/bRr18/evfuTXZ2Nq+//jozZszg8ccfj/h93HbbbSQlJXHllVeyatUqkpKSuPXWW/nXv/5FUlISkydPjupxq8jJ9KSY5T8vjkIcfwVygJZAd+By4A4zqwe8CjwHNPHrfNVfLiIiIiIiIsKUKVPo168fw4YNIycnB4CcnByGDh1Kv379mDJlSlTry8nJ4aWXXmLatGm0bdsWM6NDhw506NCBkpISZs2axaRJk2jVqhVxcXE8+eSTrFmzhpUrV5bu44YbbqBnz57ExMQwatQo9u3bx7p16yqse+rUqSQmJtKoUSNef/116tWrx0MPPUT9+vXp0qUL9913H88++2zE7+Xpp58mLy+PSy65hAULFpCXl0erVq3IzMwkLy+PBx988ISO0Yk64SSFc26ic26kc65zFOJoC7zknCtyzmUD/wbOAnoBdYAnnXOHnHN/AgzoE4U6RURERERE5DQwe/ZsHnroIebOnUvnzp3p06cPnTt35qWXXmLChAnMmTMnqvVlZWUB0KlTp+PW5ebmcujQIdq2bVu6LD4+ntTUVLZu3Vq6LNDrAiAuLg6A/Pz8cuuNiYkhLS2t9PXWrVtp06YNZla6rH379sfUE4mioiIyMzO59NJL2bJlCwcPHqRr166V2ke0VCpJYWZnmNklVRDHk8CNZtbIHz5yFd8mKj5zzrmgsp/5y8PFN8rMMs0sMzc3twrCFBERERERkdqmYcOGPPbYY3z66afEx8ezZMkS4uPj+fTTT3n00UejPodCeno6QNieDykpKdSvX780kQHe/BU5OTnHJBjKExMT/lLdzI5JSKSlpbF582aCL5k3btwYcT0vvPACSUlJJCcnU1hYyJlnnknnzp3Zu3cvSUlJ9O7dO6L9RFNESQozSzWzRcA2YJG/7AYz+2uU4ngPL/Gw368jE1gAxAP7QsruAxLC7cQ5N905l+Gcy0hJSYlSaCIiIiIiInIq6NKlCzNnzgS823h26dKlSupJTU1l8ODB3HHHHWRlZeGcY/369axfv56YmBiGDx/OhAkT2L59OwUFBYwbN47OnTtHfFeMFi1akJOTw/79+8st179/fw4dOsSUKVM4fPgwa9euZerUqYwcOTKiem666Sby8vIYPXo0kydPJi8vj6uvvpqZM2eSl5fHkiVLjil/6NAhioqKKCkp4ejRoxQVFUX9LiOR9qT4E7AJSAECEbwL9DvZAMwsBq/XxHwgDkjGm39iKnAAaByySWOg/D4wIiIiIiIi8p0U6IVQVm+EaJkxYwbdu3fn8ssvJyEhgWuvvZbs7GwAnnjiCTIyMrjwwgtp3bo1O3bsYOHChcTGxka07969e9OvXz/atm1LUlISy5YtC1suMTGRt99+m0WLFtG8eXOuuOIKhg8fzj333FOp9zJ//nwGDhxIYWEhS5Ys4Zprrglb7nvf+x4NGzbk/fff55FHHqFhw4bceuutlaqrInbsSIoyCpntBNo454rMbI9zrqm/fJ9zLvGkApuwT7QAACAASURBVDBLBnKBJOfcPn/ZAGAScA8wA0gLDPkws83Abc65f5e334yMDJeZmXkyoYmIiIiIiEg1WrNmzUn3fli6dCm9e/dmyZIlpXfFkJpR1udpZquccxnhtok0tXQIbwLL4J02BfZUNshQzrldeL00bjezOmaWBIzAm3tiKVAM/D8zq29mY/zN3j3ZekVERERERESkdok0SfE28Aczqxu07BHg9SjFMRC4Eq9HxXq8ISV3O+cOAwOA4UAe8HNggL9cRERERERE5Bjp6ek8/PDDpZNbyqmlTsVFAPgV3kSWe4EGZpYHfApcG40gnHOf4N1uNNy61cAF0ahHRERERERETm/p6elMnDixpsOQExRRksI5twf4gZllAOnAZiDTRTKhhYiIiIiIiIhIBCLtSQGAcy4T7/agIiIiIiIiIiJRFVGSwsyWAOF6TRzC61XxgnPuvWgGJiIiIiIiIiLfLZFOnBmYF2Ib8B9gq/96PdAYeNvMRlRJhCIiIiIiIiLynRDpcI8OwHXOudJbf5pZb7w7cPzEzK4FJgOzqiBGEREREREREfkOiLQnRS9gaciyZXx7R47XgNZRiUhEREREREREvpMiTVJsBQaHLBuIN/wDvCEfh6IVlIiIiIiIiIicPDPjgw8+qOkwIhZpkuJXwGwzW2pms8xsKTDHXw5wKTAz+uGJiIiIiIiIVEJWFkyc6D2fombOnEmHDh2qvd7u3buTmend0DMlJYXs7OzSdStWrKB///40b96cxMRELrjgAhYsWBD1GCJKUjjnXgfOAt4BDvrPZzvnXvPX/8s5Nz7q0YmIiIiIiIhURlYWPPLIKZ2kiMSRI0eiur+DBw+yYcMGunXrxrp164iLi6NFixal6/fs2cMNN9zAF198wd69e5kwYQJDhgzho48+imockfakwDm3wTk32Tl3h/+8IaqRiIiIiIiIiJwiDhw4wL333ku7du1ISEiga9euvP/++wAUFBQwduxY0tLSSE5OZsCAAWzZsqV02169ejFu3DgGDRpEQkIC7du359VXXwVg+fLljB49mo0bNxIfH098fDxLly5l6dKl1KlThzlz5tCuXTuaNm0KwGeffUafPn1o0qQJ7dq1Y9KkSRQXF1f6/WRmZnLuuedSt25dli9fzsUXX3zM+quvvprhw4eTnJxMTEwMAwYMoFu3bqXvOVoivbsHZtYZb6LMFMACy51zj0Y1IhEREREREZFabuTIkWzfvp3FixeTnp7Ohg3f/o5/991388knn7BixQqSkpIYO3Ys11xzDR9//DGxsbEAzJo1i4ULFzJv3jz++Mc/MmLECLZv306PHj2YNm0akyZNYv369aX7XLp0KcXFxbzxxhusXr2aunXrsm/fPvr168eYMWN488032bhxI/3796d+/fqMHx/ZYIfbbruNuXPncujQIZxzJCUlUVhYSGxsLElJSYwfP54HH3zwuO2ys7P54osv6Nat20keyWNF1JPCzIYAnwIjgYeAa/znH0Q1GhEREREREZGTUVJy7HMVyMnJ4aWXXmLatGm0bdsWM6NDhw506NCBkpISZs2axaRJk2jVqhVxcXE8+eSTrFmzhpUrV5bu44YbbqBnz57ExMQwatQo9u3bx7p16yqse+rUqSQmJtKoUSNef/116tWrx0MPPUT9+vXp0qUL9913H88++2zE7+Xpp58mLy+PSy65hAULFpCXl0erVq3IzMwkLy8vbILi4MGDDBo0iP79+9O3b9+I64pEpMM9HgSGOecuBAr859HAx1GNRkREREREROREffkl3HKL9/ctt3ivq0CWP99Fp06djluXm5vLoUOHaNu2bemy+Ph4UlNT2bp1a+myli1blv4dFxcHQH5+frn1xsTEkJaWVvp669attGnTBrPSwQ60b9/+mHoiUVRURGZmJpdeeilbtmzh4MGDdO3aNWzZ/Px8rrrqKlJTU5k9e3al6olEpEmK1sC8kGWzgWHRDUdERERERESkkgoL4aGHoHt3OHAA+vTxnrt3hwkTvPVRlJ6eDhC250NKSgr169cvTWSAN39FTk7OMQmG8sTEhL9UN7NjEhJpaWls3rwZ51zpso0bN0ZczwsvvEBSUhLJyckUFhZy5pln0rlzZ/bu3UtSUhK9e/c+pvzu3bvp27cvZ5xxBvPmzaNevXoR1VMZkSYp8oBE/++dZtYFaArERT0iERERERERkcoYPhwmT4Ybb4SvvoLFi73nG26ASZO89VGUmprK4MGDueOOO8jKysI5x/r161m/fj0xMTEMHz6cCRMmsH37dgoKChg3bhydO3fmoosuimj/LVq0ICcnh/3795dbrn///hw6dIgpU6Zw+PBh1q5dy9SpUxk5cmRE9dx0003k5eUxevRoJk+eTF5eHldffTUzZ84kLy+PJUuWlJbNzs7m8ssvp0uXLjz//PPUqRPxFJeVEmmSYhFwnf/3S/7rlcCbVRGUiIiIiIiISMQeeAAWLYLZsyE11VuWmgpz5sA773jro2zGjBl0796dyy+/nISEBK699lqys7MBeOKJJ8jIyODCCy+kdevW7Nixg4ULF5ZOmlmR3r17069fP9q2bUtSUhLLli0LWy4xMZG3336bRYsW0bx5c6644gqGDx/OPffcU6n3Mn/+fAYOHEhhYSFLlizhmmuuOa7M008/zRdffMHLL79MYmJi6Z1HpkyZUqm6KmLB3UIi2sDrWzIEaAzMcs5Ft99MlGRkZLjMzMyaDkNEREREREQitGbNGrp06VLTYUiUlPV5mtkq51xGuG0q7J9hZnWAV4FBzrki52U1XjjZYEVEREREREREglU43MM5dxS4ADha1cGY2Y1mtsbMDprZBjO7zF/e18y+MrMCM1tiZm2qOhYRERERERERqV6RzkkxBxhTlYGYWT9gKvAzIAH4AbDRzJKB+cAEvMk6M4G5VRmLiIiIiIiIiFS/SKfjPB8Ya2Z3ApuBksAK59yPohTLI8CjzrkV/utvAMxsFPCFc26e/3oisMvMOjvnvopS3SIiIsdITk4uvb2YiBwrKyuLXbt21XQYIiJyGoo0SfGe/6gSZhYLZAALzWw90ABYAIwHzgI+DZR1zh00sw3+ciUpRES+o3bv3s2wYcPYsGED9erVo2PHjjz99NOkpKQcU+6WW25h0aJFJCcnA3D99dfz4IMPVrj/9PR0NAGzSHgZGWHnOhMRETlpESUpnHOPVHEczYG6wGDgMuAI3mSdDwHxQG5I+X14Q0KO4fe6GAXQunXrKgxXRERqmpnxq1/9il69egEwfvx47r//fv7+978fV/b+++9nzJgqHbUoIiIiIlEQ6ZwUmFk7M3vAzP7iv+5kZmdFKY7AbUz/7Jzb4ZzbBTwOXA0cwLvdabDGQH7oTpxz051zGc65jNBf0kRE5PTStGnT0gQFwCWXXMLmzZtrLiAREREROWkRJSn8SS0/BS4BhvuLU4DfRyMI59xeYBvgghf7z18A3YJiiQPa+8tFREQoKSnhb3/7Gz/5yU/Crn/88cc555xzGDBgAGvWrClzP9OnTycjI4OMjAxyc0M78YmIiIhIVYu0J8Vvgeudcz8Biv1lH+NNqBkt/wB+aWapZtYEuBt4DXgFONvMBplZA+DXwGeaNFNERAJ++ctfEh8fH3ZIx+TJk1m/fj2ff/45AwcO5Morr6S4uDjMXmDUqFFkZmaSmZl53NwWIiIiIqciM+ODDz6o6TAiFmmSor1z7t/+3w7AOVeIN49EtDwGfAR8DawBVgOTnXO5wCBgMrAXuBi4MYr1iojIKezee+9l3bp1zJ07l5iY4/9Za9WqVeny4cOHc+DAAbZt21bdYYqIiIhEbObMmXTo0KHa6+3evXvpxOEpKSlkZ2eXrlu7di2XXHIJzZo1o3HjxnTt2pXp06dHPYZI7+6x1czOds79X2CBmXUDsqIViHPuCHCH/whdtwjoHK26RETk9PDAAw+watUqXn/9derXrx+2zDfffEOrVq0AeOutt4iNjS19LSIiInKqOnLkCHXrRq/fwMGDB9mwYQPdunVj3bp1xMXF0aJFi9L1LVq0YM6cObRr147Y2Fg+++wzfvjDH5Kens6PfvSjqMURaU+KPwHzzWwoEGtmg4DngCeiFomIiEglfPHFF/zmN79h+/bt9OzZk+7du3PdddcB3q8A27dvB2DEiBGcc845dOvWjUmTJrFw4ULq1Ik0Ry8iIiKnqsJCuP56WL26avZ/4MAB7r33Xtq1a0dCQgJdu3bl/fffB6CgoICxY8eSlpZGcnIyAwYMYMuWLaXb9urVi3HjxjFo0CASEhJo3749r776KgDLly9n9OjRbNy4kfj4eOLj41m6dClLly6lTp06pYmCpk2bAvDZZ5/Rp08fmjRpQrt27Zg0aVKZQ1vLk5mZybnnnkvdunVZvnw5F1988THrExMT6dixI7GxsYA3jMTMWLt27Qkdv7JEegvSZ8wM4D4gFngEeNI5Nyeq0YiIiETorLPOwjkXdt0nn3xS+veiRYuqKyQRERGpRaZMgZdfhtGjq2b/I0eOZPv27SxevJj09HQ2bNhQuu7uu+/mk08+YcWKFSQlJTF27FiuueYaPv7449KL/FmzZrFw4ULmzZvHH//4R0aMGMH27dvp0aMH06ZNY9KkSaxfv750n0uXLqW4uJg33niD1atXU7duXfbt20e/fv0YM2YMb775Jhs3bqR///7Ur1+f8ePHR/Q+brvtNubOncuhQ4dwzpGUlERhYSGxsbEkJSUxfvx4HnzwwdLy5557LmvXruXw4cOcffbZDBkyJEpH1BPxLUidc884585xzsU75852zj0b1UhEREREREREouDLL2HqVBg2DPr2jf7+c3JyeOmll5g2bRpt27bFzOjQoQMdOnSgpKSEWbNmMWnSJFq1akVcXBxPPvkka9asYeXKlaX7uOGGG+jZsycxMTGMGjWKffv2sW7dugrrnjp1KomJiTRq1IjXX3+devXq8dBDD1G/fn26dOnCfffdx7PPRn65/vTTT5OXl8cll1zCggULyMvLo1WrVmRmZpKXl3dMggK8nhsHDhxg0aJFDBw4kLi4uMgPXAQivQXp382sZ1RrFhERERERETlJWVmwdOm3j3ffhSuugIQE+MMfqqrOLAA6dep03Lrc3FwOHTpE27ZtS5fFx8eTmprK1q1bS5e1bNmy9O/AhX5+fn659cbExJCWllb6euvWrbRp0wZ/5AMA7du3P6aeSBQVFZGZmcmll17Kli1bOHjwIF27di2zfN26denbty+5ubk8+uijlaqrIpH2pIgF3jKzr8zsPjNrWeEWIiIiIiIiIlVs5kzo3fvbR9++sG0bdO8OVXVH8fT0dICwPR9SUlKoX79+aSIDvPkrcnJyjkkwlCfcHcvg23kgAtLS0ti8efMxQ2A3btwYcT0vvPACSUlJJCcnU1hYyJlnnknnzp3Zu3cvSUlJ9O7du9ztjx49GlHvj8qIKEnhnLsFaAH8DugPbDaz18zsuqhGIyIiIiIiIlIJt9wCS5Z8+1i8GNLSvAkzc3Kqps7U1FQGDx7MHXfcQVZWFs451q9fz/r164mJiWH48OFMmDCB7du3U1BQwLhx4+jcuTMXXXRRRPtv0aIFOTk57N+/v9xy/fv359ChQ0yZMoXDhw+zdu1apk6dysiRIyOq56abbiIvL4/Ro0czefJk8vLyuPrqq5k5cyZ5eXksWbKktOxbb73Fhx9+yOHDhzly5Aivvvoqzz33HFdddVVEdUWqMnNSHHTOzXDO/QDoAhjwclSjEREREZFjFedD3jPes4iIHCc9HXr1+vbRpw/8+99w4ADce2/V1Ttjxgy6d+/O5ZdfTkJCAtdeey3Z2dkAPPHEE2RkZHDhhRfSunVrduzYwcKFC0snzaxI79696devH23btiUpKYlly5aFLZeYmMjbb7/NokWLaN68OVdccQXDhw/nnnvuqdR7mT9/PgMHDqSwsJAlS5ZwzTXXHFdm//79/PznP6dp06akpKQwceJE/vCHP0ScEImUlTUzetjCZsnAUOAWoCPwinNuaFQjipKMjAyXmZlZ02GIiMgpKiMjA/07IjWuOB923glFq6DBBdD8KYhNqOmo9P0QkSqzZs0aunTpEpV9TZgAkybBokVVM3mmVKysz9PMVjnnMsJtE+nEmT8xs/nANuAG4K9Ay9qaoBARERE55QUSFIfXQJ223vPOO9WjQkQkQg88AIMHQ9OmNR2JVEakwz2mAeuA7s65Hs656c658gfHiIiISNXIyoKJE73n2qi2x3cqCE5QxDQHM+9ZiQoRkYg1bAjz5sF559V0JFIZkSYp0pxz9znnvqrSaERERKRiWVnwyCMUrt3C9dd7E4PVKrU9vtouNEGxrRAeX+s9K1EhIiKnuUjv7lFsZhea2d/8u3r8zcwurOrgREREpGxTnm/Nyy/Dnj01HQmsXr2aRYsWHbMsEN8776xitTIVkQnXg2JbATy5jsKNxVx/+/2s/rKHEhUiInLainROigHAe0AisBpoDCzTLUhFRESqR2gS4Eu6MPXF1gwbBs4tqvEkwJQpU+jXrx/Dhg1j7969pfG1bfs+992XwZQpU2o0vlNCUIJi9ZcNWPTBrmNWT3llBC+/cSnvfFDC6i8bKFEhIiKnpUiHezwMDHLO3eScm+CcuxkYBEyssshEREQq8PXXX9OjRw86depEjx49WLdu3XFliouLufPOO2nfvj0dOnTg2WefrYFIT05WFtx11wL69ZtEv36TeXVpDFfwFvENjlBQcDv9+vWr0SRAVhb84hfPMXTos7z44g763/QsV/AWHM1j69YbmDBhAnPmzKmx+E4JIT0opjy1gX43f8iwsavZu++Il/R5dSht017mvt88xZSnNmjoh4ictkpKSmo6BImCE/0c60RYLh34d8iyt4AXT6hWERGRKBg9ejR33nknQ4cO5bnnnuO2227j3XffPabM888/z/r161m3bh27d+/mvPPO44c//CHp6ek1E/QJmDkT3nvvEcC7jVqgQ8X3Dyxm3avPMmHCBB544IEaje+RR+oDI4GRLC/24+NdZs1/mvZh7rUuQUISFFnbmvOLIcNo0OAbXnz1Gza8WYetvALF+9m64w4mjO3IA3d2OH4yzVpye1IRkZMRFxfHN998Q/Pmzalbty5mVtMhSSU55zhy5Ag7d+4kLi6u0ttHmqTYDPwQeDtoWV9gS6VrFBERiYKcnBw+/vhj3nnnHQCGDBnCmDFjyM3NJSUlpbTc3LlzufXWW4mJiSElJYUBAwYwb948xo8fX1OhV9ott0CvHofguedwL7xAvjN+4WbwBefxlUuhuXPgXK2Ijxdf5HC9Bgwt/Ctfch4JA8+G+1d694Fr2LDGYqzV8v8JRau824yaMXNeXx558qbS1d8mfRYz6+k2tO/X4tttA4mKolXefpJurebgRUSi68wzz2TXrl1s3ryZo0eP1nQ4coLq1KlDYmIiycnJld82wnKPAa+a2cvAJryeFYOAEZWuUUREJAq2bt1Kq1atiI2NBSA2NpYzzjiDrVu3HpOk2LJlC23atCl93bp1a7Zu3Rp2n9OnT2f69OkA5ObmVmH0lZOeDunjh8LLL8OwYbx/7bXkDr6SOnX+j/FpLzF70mXw1VfefdZqQXz/ue46cgd68d2bNrfG46v1Em6EgvdLe1Lccv1iel3wKczfBgu3c7huHYYW/d1L+ty2Be7YB2M6QINYLzlVshMaXODtR0TkFBcTE0Nqaiqpqak1HYrUkEjv7vE/eD0nCoAMoBDo55x7uQpjExERqVajRo0iMzOTzMzMYxIdtcIDD3jjPGbPprhZM2ANQ4ZsYc6mS1k8NdNbX0viO9KkSe2LrzaLTfCGatTrAiU7ST9zJ71enEOvV16g17XriPvTXnK5kvzYBO5t+RT8aR3c9cm3CYp6XTTUQ0REThuR9qTAOfe/wP9WYSwiIiIRS0tL45tvvqG4uJjY2FiKi4vZvn07aWlpx5Rr3bo1mzdv5sILvTtnh/asOGWcd95xi26+eQuFhek07XcBHL+6etX2+Gq7QKIiMDfFne1haBu4NJkjy3cBaxhy7XPMmf8zRvzXnfS99FMlKERE5LQU6S1IHzOzniHLvm9mj0QzGDPraGZFZvZc0LKbzGyzmR00swVm1jSadYqIyKkpNTWV7t278+KL3hzOL774Iuedd95xPSCuv/56nnnmGUpKSsjNzWXBggUMHjy4JkKOuvr1S5g3L2x+oEalp6fz8MMP873vta6V8dVawT0quhbB95sds/rm62Yx+OoPaHppjLdeCQoRETkNRXoL0pHAZyHLPgN+Ed1weAr4KPDCzM4CngaGAc3xhpv8Ncp1iojIKWratGn8+c9/plOnTvz5z39m2rRpAFx99dVkZmYCMGzYMNq1a0fHjh255JJL+PWvf03btm1rMuyTFkgC1NY7lKSnpzNx4sRaG1+tFjL0A+dIP7MRD9/Vke+1i2Xe337LeV2XK0EhIiKnLXMRzAZuZnuBZs65kqBlscAe51xiVAIxuxEYCHwJdHDODTWzKUC6c+4mv0x7YI0fS7k3BM/IyHCB/6CKiIhUVkZGBvp3RGpMyG1JMatVc1Do+yEiIifDzFY55zLCrYu0J8V64IqQZT8ENpxMYAFm1hh4FLgnZNVZwKeBF865DcBhoFM06hURERGplcL0qKgtCQoREZGqFOnEmb8B5prZ34CvgY7A7XjDQKLhMeDvzrltZha8PB7YF1J2HxD2X2YzGwWMAm+iNBEREZFTVvBkmkWrvNuMKkEhIiKnuYiSFM65+WZWCIwBfgxkAUOcc2+cbABm1h2vV0a4abUOAI1DljUGwg71cM5NB6aDN9zjZGMTEZHvps8//7ymQxDxBBIV+f+EhBtrTYKisLCwpkMQEZHTVGVuQfom8GYVxNALSAe2+L0o4oFYM+sK/BvoFihoZu2A+ni9OURERKpEUVERX331FRkZYYdK1gq5ubnH3cmkNlF8Jyd8fE9XexzOOY4ePcrRo0dLb/dbp04dNm/eXO2xiIjId0PESYoqNB34Z9Dre/GSFrcDqcByM7sM+Bhv3or5FU2aKSIicjLq1atHhw4davXEgLV94kLFd3JqMr6CggJ27drFrl27KCgooGnTpiQnJ9O0aVPq1KlTGp+IiEhVqPEkhXOuAO/WogCY2QGgyDmXC+Sa2WjgeaAZsAj4WY0EKiIi3xn169cnkrtfiZwOnHPk5+eXJiaOHj1KcnIy6enpJCUlERMT6TzrIiIiJ6/GkxShnHMTQ16/ALxQM9GIiMh3kZIUcrorKSlh37595Obmsnv3bmJiYkhJSaFz584kJCQQMpG5iIhItal1SQoREZGaVq9ePa6//vqaDqNco0aNqukQyqX4Tk5VxFdcXMyePXvYtWsXu3fvpmHDhqSkpNCtWzcaNWpU4/GJiIgAWCS/FJnZGudclzDLP3fOnVMlkZ2kjIwMV5vHmoqISO21a9cuduzYwTnn1Mp/4kQidvjwYXbv3s2uXbvIy8ujcePGJCcnk5ycTP369Ws6PBER+Y4ys1XOubATHEXak+LMSi4XERE5ZdWrV49Dhw7VdBgiJ6SwsLB0fomDBw/SpEkTUlNT6dKlS+nElyIiIrVVuTMhmdkDZvYAUCfwd9BjBrC1esIUERGJnqKiIm6//XY6duzIOeecU9p1PT09nc6dO3PZZZcxZMgQ3nrrrdJtVqxYQbdu3ejUqRM/+tGPyMnJqfb4yloO8PXXX9OjRw86depEjx49WLduXbXGl5WVRffu3Usf6enpNG3atNbEB/Daa69x3nnn0b17d7p168b8+fNrJL7yYnz99dc5//zzOeecc7j88svZtGlThTEGJr7ctGkTH330ER9//DEFBQW0bt2anj17ctZZZ9G8efOIExTlfZblHafqPoYiInKacs6V+QCW+I+jQX8vARbj3XEjo7zta/JxwQUXOBERkXB++ctfurvuusuVlJQ455zLzs52zjnXpk0b9/nnn7uSkhK3dOnS0vXFxcWuffv27v3333fOOffYY4+5n/3sZ9UeX1nLnXOud+/ebs6cOc455+bMmeN69+5d7fEFGzt2rLvzzjtrTXwlJSUuKSnJff7558455z799FMXHx/viouLqz2+smLcs2ePa9asmVu7dm1pHFdccUXpNsExzp4921122WVu3bp1bvny5W758uVu/fr1Li8vr3Sf0RL8WZZ3nKr7GIqIyKkLyHRl5SHKWnFMIfhzJOVq00NJChERCSc/P98lJia6/Pz849YFkhTOOfef//zHFRUVOeecW7lypTvrrLNKy+Xm5rq4uLhqja+8uHfu3OkSExPd0aNHnXPOHT161CUmJrqcnJxqiy/YoUOHXHJyslu1alWtia+kpMQ1bdrUffDBB84555YtW+Y6duxY7fGVF+PKlStd165dS1/v3r3bAS43N7c0xuzsbLdmzRq3bNkyFx8f7zIzM11+fn7UExMBwZ9lecepuo+hiIic2spLUkR042vn3C+j3oVDRESkBmzYsIFmzZrxyCOPkJGRQa9evfjggw9K1998882ce+65PPHEE+zcuROALVu20KZNm9IyycnJlJSUsGfPnmqLr7y4t27dSqtWrYiNjQUgNjaWM844g61boz8qs6LjB7Bw4UJatWrF+eefX2viMzNeeuklrr32Wtq0acOAAQOYPXt2tcdXXoydOnUiOzubjz76CIDnn38egI8//pglS5bQpEkTduzYQXx8PBdddBFpaWmYGfHx8VV2y9Dgz7K841Tdx1BE/n97dx4fVXk1cPx3skIWSMKwk4gICEgRlFdRUJDFKmBVBDegWLXU1q21Lq1aBWupWK3WVqVUrQgVF8StWi3EYEWrryzVVwStmAQBicRA2ELW8/5xZyaTycxkkkwyk3C+n898ktz1zPPc89xnntx73axAkQAAIABJREFUx5j2K6xBChHpLiJ/FpH1IvK576ulAzTGGGMiqbq6mi+//JIRI0awbt06Fi5cyLRp09i3bx/vvPMOH330ER9++CEiws9//vOYia+0tDRo3LEQn28cTzzxBJdffnmrxhVOfL/97W95+eWXKSws5NVXX+XCCy/kwIEDMROjiPDss89y/fXXM2zYMD7++GPS0tLYv38/GRkZdOzYkeHDh9OnTx86dOjQKrFGsy6NMcYcmcJ9xPNTQCrwOHCw5cIxxhhjWlZOTg4JCQlccsklAJx88sm4XC4+//xzRo50vgkrOTmZ2bNnc/XVV3vXKSws9G6juLiYuLi4Og+GbOn4UlJSgsadk5PDjh07qK6uJj4+nurqanbu3El2dnarxecpvx07dvD222+zdOlS7zrZ2dkxEd/OnTsZPXo0AKNHjyY1NZXNmzdz1FFHtVp8gWI86aSTyMrK4u2336Znz5787ne/o0uXLtTU1LB06VLOOussDh48yM6dO1stRqBeXYaqR1Vt1TI0xhjTfoV1JQUwCjhLVR9R1SW+r5YMzhhjjIk0l8vFGWecwapVqwDnGwm++eYbevbsSWlpKeA8r+mNN95g0KBBAJx44omUlZV5b2tYtGgRM2bMaNX4+vfvH3R6t27dGD58OMuXLwdg+fLljBgxgq5du7ZqfABLlixhypQpdOnSxbtOLMTXp08ftm/fzmeffQbA5s2bKSoq4phjjmnV+HxjfOmll9i6dSsrVqxg165d9OrVi/T0dE499VQGDhzI/fffz1VXXUVqamqrxwj16zJUDNGIzxhjTDsV7GEVvi9gE5AZzrKx8rIHZxpjjAlm69atOnbsWB06dKiOGDFCX3/9dd26dasOHz5cv/Od7+iQIUN06tSp+vbbb3vXeffdd3Xo0KHav39/nThxYsBvtGjJ+EJNV1XdvHmznnTSSTpgwAA96aSTdMuWLa0en6rqgAED9B//+Ee9dWIhvmXLlunQoUN12LBhOmzYMH3xxRdbNb7q6motLi7WLVu26HPPPacnnHCCDhw4UI8//nh97bXXVFX1iiuu0EGDBmm/fv30qquu0rKyslaN0VegugwVQ2vHZ4wxpu0ixIMzxZkfmojMAi4E5gG7/AY5dkZ85CQCRo4cqevWrYt2GMYYY9qokpISvvrqK44//vhoh2LasKqqKr799luKi4vZs2cPqampuFwuXC4XHTt2jHZ4xhhjTFSIyHpVHRloXmOeSQEwFfCMaoj79/jmhWeMMcbEnqSkJCoqKqIdhmmDysvL+fbbb9m9ezf79u0jIyMDl8vFgAEDSEpKinZ4xhhjTEwLd5Di6BaNwhhjjIkxycnJlJeXRzsM00YcOnSI4uJiiouLOXToEF26dKFXr14cd9xxJCSE290yxhhjTFhnTVUtbHgpY4wxpv1ISEigpqbG+20FxvhSVfbv3+8dmKiqqsLlctG3b18yMjKIiwv32eTGGGOM8RX20L6ITAImAF1xbvUAQFXty7ONMca0OyLiveXDnh1gAGpqati7d693YCIhIQGXy8WgQYNIT09HRBreiDHGGGNCCmuQQkSuB34LvIbzXIq/A2cDK1suNGOMMSa6PLd82CDFkauqqoqSkhKKi4spKSkhJSUFl8vF8OHDSUlJiXZ4xhhjTLsT7pUU1wCTVXWNiOxR1RkiMgWY1oKxGWOMMVFlD888MlVUVHi/kWPv3r107twZl8vFMcccQ3JycrTDM8YYY9q1cAcpeqjqGvfvnm/3eB1YAlwR6aCMMcaYWGAPzzxylJWVeW/jOHjwIJmZmXTv3p3Bgwfbgy+NMcaYVhTuWfcbEemuqkXAdhE5GSgGIvJUKBFJBh4BJgJZwFbgl6r6D/f8CcDDQA7wAXCZPczTGGNMS0tKSrJBinZKVTlw4IB3YKKyspIuXbqQk5NDZmamPfjSGGOMiZJwBymewXlo5tPAY0AeUAU8FcE4vgLGAtuAycBzIvId4ADOsy+uBF4Ffg08C4yK0L6NMcaYgJKTkzlw4EC0wzARoqp1HnwZFxeHy+Vi4MCBdOrUyR58aYwxxsSAcL+C9Daf3x8SkfVAOvBmJIJQ1YPAPJ9JfxeRfOBEoAuwSVWfBxCReUCxiAxS1S2R2L8xxhgTiN3u0fZVV1ezZ88edu/eTUlJCR06dMDlcjFs2DBSUlJsYMIYY4yJMU26yVJV3410IL5EpDswENgE/Bj4yGffB0VkK3AcsMVvvbnAXICcnJyWDNEYY8wRwB6c2TZVVlZ6H3y5Z88e0tPT6dq1K0cffTQdOnSIdnjGGGOMCSHoIIWIPKSq17l/XxxsOVWdG8mARCQR+BuwRFW3iEgasNtvsVKcKzn8Y1kMLAYYOXKk+s83xhhjGsNzJYWq2n/cY9zhw4e9t3Hs37+fzMxMXC4Xxx57LImJidEOzxhjjDFhCnUlRWKQ31uMiMQBS4EKnK89BeeZFJ38Fu0E7G+NmIwxxhy54uPjERGqq6vtGx5ijKpy8OBB78BEeXk5Xbp0oU+fPmRmZhIfHx/tEI0xxhjTBEF7XKr6YwBx/nV0C1CsqjUtFYh7P48D3YHJqlrpnrUJmOOzXCpwjHu6McYY06I8V1PYIEX0qSqlpaXegQkAl8tF//796dy5s13tYowxxrQD4fS4BCjEub2ixQYpgEeBwcBEVS3zmf4i8DsRuQB4DbgD+NgemmmMMaY1eAYpUlNTox3KEammpoY9e/Z4ByaSk5NxuVwMHTqU1NRUG5gwxhhj2pkGBylUtUZEvgQyqf9siIgQkaOAHwHlwC6fDsePVPVv7gGKPwHLgA+Ai1siDmOMMcafPTyz9VVVVXkffFlSUkJaWhoul4sTTjiBjh07Rjs8Y4wxxrSgcK9dvR942v31n4X4XFGhqjubG4SqFuJcsRFs/mpgUHP3Y4wxxjSWfQ1p6ygvL/deLbFv3z4yMjJwuVwMGDCApKSkaIdnjDHGmFYS7iDFY+6fEwDPt2aI+3d7MpUxxph2KykpibKysoYXNI3m++DLsrIyunTpQq9evRg6dKg9+NIYY4w5QoU7SHF0i0ZhjDHGxKjk5GT27t0b7TDaBVVl//793oGJ6upqXC4X/fr1o3PnzsTFxUU7RGOMMcZEWViDFO7bMYwxxpgjTlJSkt3u0Qw1NTXs3bvXOzCRmJiIy+Vi8ODBpKWl2YMvjTHGGFNH2N+nJiKTcG736IrP8yNU9fIWiMsYY4yJCcnJyfbgzEaqqqqipKTE++DLlJQUXC4Xw4cPJyUlJdrhGWOMMSaGhTVIISLXA7/F+QrQqcDfgbOBlS0XmjHGGBN9nm/3UFX7r38IFRUV3qslSktL6dy5My6Xi/79+9uDL40xxhgTtnCvpLgGmKyqa0Rkj6rOEJEpwLQWjM0YY4yJuri4OBISEqisrLQP237KysooLi5m9+7dHDp0iKysLHr06MGQIUNISAj7Yk1jjDHGGK9wexA9VHWN+3fPt3u8DiwBroh0UMYYY0ws8XwN6ZE+SKGqHDhwwHvFRGVlJS6Xi759+5KRkWEPvjTGGGNMs4U7SPGNiHRX1SJgu4icDBQD1hsxxhjT7nkenpmenh7tUFpdTU0NpaWl3oGJuLg4XC4Xxx57LOnp6XYLjDHGGGMiKtxBimdwHpr5NPAYkAdUAU+1UFzGGGNMzDjSHp5ZXV3tffDlt99+S8eOHXG5XAwbNoyUlBQbmDDGGGNMiwk5SCEiDwGLVPU2zzRVfUhE1gGdgDdbOD5jjDEm6jy3e7RnlZWV3qsl9u7dS6dOnXC5XPTr14/k5ORoh2eMMcaYI0RDV1J8B/hYRD4AFgHPqWq5qr7X8qEZY4wxsSEpKYn9+/dHO4yIO3z4sHdgYv/+/WRlZdGtWzcGDRpEYmJitMMzxhhjzBEo5CCFqp4hIgOAucB9wIMisgzn6orNrRGgMaaFVO+H/c9A+sUQf+TdZ29MYyQnJ1NcXBztMJpNVTl48KB3YKK8vByXy0V2djYZGRnEx8dHO0RjjDHGHOEafPClqv5XVW8C+gA/AY4DPhGRd0RkVksHaIxpAdX7oehqKHnQ+VndBv9DXL0f9v6lbcZu2pykpKQ2+0wKVWXv3r188cUXfPDBB3zyySdUVVXRv39/Tj31VI499li6dOnS+gMUlsMmFtlxGR4rJ9MUdtyYMIX97RyqWqmqz6rqROBEoBfOV5AaY5qqKY11cxt4zwBFxWZIONr5GYmBitY88bSHQRbTpkT8mRSRypcg26murqa4uJgtW7bw3nvv8cUXX5CQkMDQoUM5eeQQ+rvyyEiPj94DMGMlh2OhwxwLMbQFrVFOsXJcxrpYLyfLqdgU68eNiSmN+gpREfkfEfkL8C/3ur9qkaiMORI0pbFubgPvO0AR1x1EnJ/NHahozRNPSw2yGBNCYmIiVVVV1NTUNH9jkcoXv+1UHi6hqKiITz75hPfee4/t27eTlpbGiSeeyMiRI+nbty9pHRX55prodhJjJYdjocMcCzG0Ba1RTrFyXMa6WC8ny6nYFOvHjYk5DQ5SiEgnEblaRP4DrAW6ABcC/VR1QUsHaEy71JTGurkNfKABCmj+QEVrnnhaapDFmAaISGRu+YhUvri3U37gv+zYM5SPNu/j/bfmsbtoGy6Xi1GjRjF8+HD69OlDhw4dIrvv5oiVHI61srBOe3CtUU6xclzGulgvJ8up2BTrx42JSSEHKURkCbATuAlYAfRV1Wmq+oaqamsEeMQqKIB585yfR7JA5RBu2QRbrillW1AAP/2p8yooaF79NKWx9l1nRzo88LnzM9wGPtgAhYd/DFs/Ce/9NTauhsot1PxwBlnWzYE7ftn0Y6MtaQ/voY1JTk6mfOvWprcrzc1j3A++3FdE4X9+zvpPDvLhhwPZ9+xuelVXceqx7zPU9Ud6dE2t/80cEdh3s7XUQGlz4oiFsohWDI3VGm2O/z5a48ON/z62l8HvP3N+BtrXkdr2NracWkK4fYS2klO+mtPfbcq2W3I9X8Ha/e1l8IdS+PI/4dVPS+SdZ5tr17bNnPYtk3bYLjV0JUVnYAZwtKrerapft0JMR7SyMpgxAzauKob58yn7bJvz98YoxbExvOnN3W5QBQX1yyHQtHDX9Zne6EGKP/yBsj/8mWmzU5h0QSc2zn85ZAwB32uQk3zZ1mpmXP0AGzdW12+sfdYpq8hmxs9vZ+ODqZTl1wRfh/r7LNv3pbP8pmMCL+fb+dt4Q8Nl7P9edpTBg/8NHVdDdRdsfriDLPmb4Nf3wNZP6zfY/o25735e2wnz5lG2pbBl8q2lTq7hHMeR3Hcjt9Wk9iKcfUTpZJyUlERFA+1K0GO7KfnipqqUlpaydetW/veDd/j433dTUbadfr2VUxM+Z/Dv8kj75gAXX/f7BtuQxu47YrZ+AjePcTqk4QyUtlQsrVkWwY7TWKiPpgj33BupfUyvZGPuvfXa/bLy5MDl1JR2IdC5Zfshpz6+rGba3FuZNGtx3X0FaHub2zeKeX7lVFaezIwbbnP6Il9WM+PHv2Djp6e0fP6G20dw51TJZwn0OeUfLF/RJ6y4olqPzenvNmXbLbmeR6i+myfP8rOYceX3nVwPVT8t0f54cnndusZ/LmhhYR2Lvm1Ra7TPrSzkIIWqnqeq/4j2VRMikiUiL4rIQREpFJFLoxlPpGzcuJHVq1fXmbZgAaxYAe+vz3f+/lsOK1bAqlXrA25j9erVbHQfiYG2F2i5xsSxatX6Out5pn/wwX+btC/P+iUlAWcHdeMf0+qt51s2ocog0HL+AjUGGzduZNGiRXW2t4BbeXFtN1ZvyKKELG9cS5a8ykMP/YVp02DSJGc79crQ3Vhv3Lie1e/F1+kMLfjThax4fQyr3s1k48b1tSdTv3UWPHwhKz4YTwlZLHhxTuB1fPmcIBY8cg0rXh9Dyd604AXt+aBQVei836U9A9dXoPfiKSO/uDbmXsLqN18JWifLly8PXWfr1nrfw8ZPO7B6bd2vgSw7nMSMq25h0bIs/luQ5Ezc86AzUDF/PutfeAGA/65aVduAzxvCRoZ76+/9V9+F+fNZsECbdHw2qAVPHmXlcWEN1kVskCLIthrTjgSzceNGp74CfAA4/fQili/fUieO1j4ZJycnU15dDQRvmwO2N43IF08e19TUUFJSwmeffca///1vPv/8c+I4zBDXX0mufIvCbSVkdqokTkJvpyn79uXfNjbpPFO93xn4/P3HsCO1/gCFR0sPVDSzLBotUL60dgzN1NhzalOOj4D7WNqTFS8ksmT5Vhb97XDdcvrTjMDl1Nh2wef8uPHTDixaVsjqd3bX7ufFObz45qmsfncES17o6LOvQ/U21dS+TZvgV06r1xY7deDfF1lbw8ZPO0Qsfz3HRaDj49r7Euv1EZavWMu9iw/UOVbOvfdeduxyseylDJavWNtgXJGox8bkQEP59d///rfOtOYeX+Fsp6n96HoCDFBs/KS0To4BLHhpDiv+MZFVedvYmHtJnfqJWCwxLlT/6dFHnwn4PlevXu09PuqsF6FjJSaoasy/gOXAs0AaMAYoBY4Ltc6JJ56osW769OkK6KxZs7SoqEg3bVJNTKzRo4/+l44F3cRgTUyo1pycNXWWU1UtKirSmTNnKqDTp08PuL1gyzUmDt/1nOmqs2c3bV++64ctL083MVjjpEJhic6aNUtLVq70lo1/jL5xBV0uL08VnJ9ut9/uTFq9un65AHrH6afrJgZrAuUKVdqdJd64PNuGuxSc7Tz+uF8ZXnCe6o7ZqvkjdfrkHk6M5/fWksUnOjHGV+jR2c87y07uoZo/UnX7xc7Lu85gjZNyvWjU34Ovs2O2atU+3bBhg65642XvPjetukoTEyt09rRcXfW3k3XD66epbpsa/PXMKO8+pkwu0lWrVtUWTNU+73ZPOymrwfcy/exMZ5mZFwWsk969e4eusyk5znsrnKLTJ/f07q9owyTVbVP1Z1cuc5f7eL3j5CynApYPVH1+kiroWPe2f3nKKaqgp+U86dQRP9A4Kurl2+zZqqtWrdINGzaoat3fG2PDhg215eY+5m6fna+gunDhuiZtsw6/bfoeu4H27Tnem/p+Am3LV7jtSCjTp0/Xse4kKlm5UlWddmXo0BcVVIcOvd6JPcR7bwme8iwsLNQvXnihXts8efLk4O2NT740lPszpvTSog1n6KZ3rtJ33l6tGzZs0G3btumhQ4catZ3gbUgY67jbEA//trHRbb8n7uUDnQ09Oyp027NtqmrhlICxNEtTyq+5+/fPl2jE0ExhnVODLB9u3yDgPuIr9Og+z3jPwZ42f9PqH2tiQm05TRjdRVctHeiU0+rXwm9nferC99wC6B0nZznn+7gKhWqFJXXrxH1u8dRrk/o2bUXAcgrVF+kZsfz1HBc5OTne4+PLJ57QTQxWoVxTUlbU6SP07pGsgB6d3VHX/GyAvsxUhRpNiN/q9DV6JKvmj9RVyyfqhg/fqbe/SNVjY3Kgofzy9D09fZPG8O8HBOvjhIo/nJwPyO+48bTtvn24gO3e2Zl1jpuIxBKKp41+4IGgfZvWEKz/5DnGc3JyAh5Ld5x+em3cfnXcVgDrNNjn/2AzYuUFpAIVwECfaUuBe0KtF+uDFPn5qm+8cVhnzXpM4+MnaErKlZqYWKJQqnFx0/SmiQu0D9s0K71CCwsP6e23366JiYmamZmpZ5xxhmZmZmpiYqL+6le/0rKyMlVVPXQovOVCxZGW9j1NTi5S2KNxcdN09uzH9c03yzU3V7VPH9WsLNVvvml4X1u2HPbkjOblab31Q5VLnfXu26A92KHpHSt0+vRlGh8/QU/pMFn7sE0T+VYTEnp635v/ewm2nObmOod+bq6qBj4x5eerPvnkYR0//hWFn2oO87UTJSpUaRylehznebcdH/8/OnnycwpVClWakJCryclFKuLe7203a9nWS1XzR2r+2h/oG0tv1lnTZmp8/Dgd1WGiezvFmpDQVX91/QAt+/xszX/n+5r31Pc076nv6fKHFug9v/izduxQrLBXRyZPC7iOp2OQ//7P9PTTb1MYqxNHT9EVj96ofXp+o5mdS/WCyd+p7Uj4fCjIf/dyzXv2F95X7m1XO8dfWomec9b42sZz51bVHbO16MPj9dxJJymM1exeZ2l8/Dg9JcB72fLWHH3jqZt01nnTND7+jHp1cu219+srr+wLXmc//Y6WbTnBe5I79PnZevt1AzQxUTSzc6KeNPwUhXIVWaK/un6Ali85yanbp0/2fih6bOZFmpDQX0d1OEv/yvcVKrQTb6uLIo1jr8K5OsE1Q/uwTTsm7NWpU3+ggE6dOrXBAb5QWvzkmpsb9ITkv2/Ph/7mvB9VDTpIEawd8eZAkPbH36FDh/T7OTmqoBMTEnT06NHaqdMohXLNyHi5NvZWPBnn56uefvodCmP13HN/rU/euUL7sE07pRzW0aPvVRirMC5we3PgG29HzT/3PfmSmrRTs7OH6B03TdK3XrxCP35rtu5cN17LC75f28H36fAF20699qDgu6qfdVH9rIvmr/1+eOv4fbgI1DY26jzj21F9ZlT4gxSRHqhoSvlFYv+++RKtGJoh7HOqj8b2Q4Lvo1jj40/QS867VC8591KNjx+naamTNDlpu4rUltP5Z7kHe87L0pLHTg+vnfX7AJX/7uX65O9/q+NH/1rhej2Gn6qLIvc5/VsVcekvrz7GOTev/YHm3TZL8xireb97L+y+TZtUtU/z3/+Z5i2fq3nP/EKX//Ge8PoiEcrfQ4cO6TXX3KdxceM1MXGSwjgdzhk++6ztI7zz/I16902PaN/sX7rr8AaNp9I9yHSsnnZSln6+ZqzOPK+XAnr2Gadq3uqDje6jhht3ODkQ6NjvwQ6Nd38OmD37cX3jtx94Pws0Ji7f89bEiXfrS3e+rX3Ypplph/WCC64KmhtNyfl6ggxQ5L97ecPt3mdneY+b/C/2Nz+Whng+D9x3X53PBa0pVP8pPr6HjhkzJuCxdO219+s/F37gtEX3r9fc+zY06ViJtrY+SDECOOQ37Ubg1VDrxfogxZ13OqXf0Gt88lrnU7Sqfvrpp5qdna2AZmdn66effhpw2+EuFyqO+PjCwPGMD29fwbbrv35TywVUR5OrX7zySoPr+i6X/8/PNa/rDCepu87Q3MfzA56Ywo2jL5/pmOFFAeclJ6/VT//vwzqN9Z0//VvAZbsl5OqnuWPdJ/ez9M5r7g/j/a/WLx4fWa9jf+f1fw64fEJ8riYmSt2OhPsVLK7xSXla9uG/ak+4GUl6xinpmtk5UePi5jUYV7DteuokZJ0tv1p16xDvSc53IOXJ+6/Xrl2+q7BN46RY177wPdXVY1V7dXA20KuD5v/iu0493/2wnnvunrDqM4PVCmjfvn0bHOALJSIn+gDb9A7g/TVfc7te5JyQ4kr0m3e21FnWt5P0vbQ0VdDvpaU1+f14+Q3weQSrx+TktSHbn0Dv8c173tc8xupwxiqcofCV1hs0bcWTse9769SpVM8770UdNy5PvzdupY4bl1fvNXvcE/rCokWal/uG5r1ynea9eKnmvfRDvfMXD9Vb9swz39CrL/mDvvXHqVr00XStzD+nfge/YkdYbUid9qDwLNXPe6tu6ay6pZPeeW3g9sR3HW9+PfMLzVs+V3OfXaB9+lQH7bSHdZ7Zs7g2h5e7BymWhzlI4SmHrUOc7TSVX4c5rPKL1ECJJ1/++Wr0YmiGcM6pwYTbDwm+j9V67fcCn8uSk97ST3PHBvjQM6HuB9hgbZ3vcRniPAWqGb7n5m1T9c7Zfw18rmygb9Mm7Vmsd173p6b1RSKUvyGPwSfP8ObU2JM/DrhcZ77VEzvG6RmndtHMzolOH+i6/nrb1YtavB4byoHG9Hd9Pws0r9xW67CEhKC50Zyc9/LLr4b6mfWOH/dxc+ct65ofSyibNqlmZzsbjY93fmZnN6qcIyGc/lOgY+nOn3wTkWMl2kINUogzP3aJyGnA86raw2faD4GZqjrOb9m5wFyAnJycEwsLC1sz1EbxPOPEY9WqTSxY4CI1oRNLa2bTOaWSyyoXc6AiiS3xQ+n2i8vh1lt569//ZsKECeTm5jJ+/Pig23/rrbfCWs4/jvXrN3LjjV1ITe3J0qWJZGY602tq4LLL4MAB2LIFunULvS//7YZav148n5XDsmWwfDk1HVOZVfEY+ys68BRzyJS9VCR3ZNbhR6kinS0JtWVTUNSxzroVSR2YVfaIs1z8cXQb1Y95//4u82t+VW+/48dVk5sXXyeO998uZ9eSNyEvj91k8jDXUEpnEqmikuSA8Y8YATNnOmWYnt6TL9Y/Q7f4e5yvwhKh4KtuFGx3v/mvyjh812bm7vsLBySNLc9dTreTq6GykIL8agp2DgCEXd90Yuc3x3HPQ5dTdjiJx5Mv55ry31NFGlsSBtPtJ53hmv7QwYm/YFtXCgproNN0CncP4KabfsbubxcRJ2m8/fzljDmpun65f9WNgi9dsHI7vLKTmg4JXFb5JAcqU7zH35aJcOYl9/DV11Vk9+rAE787l4Sao7zrVCQmMOvw43XiKjjvVAp2dPdZJp5Zh5/w1t2hq26gYNIP4fnn69dZwlC6Xd0FrkqElJ7Me+BS5j9Y/3E0I4Z8yIYJl8KirZCSAMd1go9LmXfgNuYzr/7yrOdnSX/k2oqFVJPCEplDkpbyE/7KPtIYnDWa90u2kJ2dzZtvvsngwYMDH6whzJvn3IrubzRvseSVgxxzzjkR2+b4hLfJlUlwyy1w663QsaN33tZXX6XD+efTu7qaHfHxHH7xxSbtG4BPP4WzzoKvvoLsbHjjDRgyBAjejqSn9+SLLxKD5nu993j1buY/0rXe9Pj4bVRX59SbPj75XXI3ZHrjaAm+bZI+/TRfnNSblzpcQBXJnMfLdOzWCXbvpjo+gZVVU6ghiWviHiV1VA9ox5wgAAAbc0lEQVQYvgeSMgFh775U9u5Jgc374LP9VMfHsbLqe5QdTuOZdZfWy2NUoSofOpwIh9fXbUN8cjVQ3vGjJIgvBOkICAVfuSjYfhK8XBF0nXk1dzL/TzPrl/F4yM0NXDYNnmc89yRv2gCXbYWdh6FXB3jqZBiYHrrgVaGmCJIGQ/eHIb6B5YPZ+xcoebBx5ReoHrJ+Chk/DH+/vvnSOxMez4LBg1o3hmbyPx/Xa5/d517fNsdXOP2Quvt4mopEcZ8j0vhn/JkcOHcQnN8bkuJY//F+bvzNzaSnZvDFOz/hkacmBzwnjCaXJa8cCt7W+d0rX7C9O+9/MIBdK6vgvWIq4+P5TeXN7MNFZ/bx2ZVn0+26zvCXLyl4tIqC5H5wdGcoqKCmrILLUp/jQHxntmyRsNu6NqF6PwXr7qQg/yBIFrt2Z7BzRyb3PDit4b5IhPLXc3zsuu8+uqxeRVVSR+ZU/Nm9z0F0+3FnuHYAaz8eyroNx8C/dsN7xZRJArdX34MSxwaG8iqbeapnMq8sG8XgAel1+kmkTQ27j9pYoXIgUH5dUvZnKkjnKS4jU/ZSk5oe8LNAsJwLtG19+mn2q3ClPkElaWyJH0r3X14RcDvNzXkg6MMyw2r3rj4GkoohaTAFZY9Q8EVi82IJpKwMfvMbWLgQEhKgvLx2XnIyVFcH7E+1lHD7T55jKe/11xn37rsULHyWguRjIS0NinZRQxyXxS/lQE1Ko46VaBOR9ao6MuDMYKMXsfIi8JUUP6eNX0nhLy8vT3/PYE2kXGf3e0fV556k2Ue/4wyPTZ+ueXl5CmheA/dNhbtc4PUGa0KAy6iD3a8X7r7Cvt9v+nTn/c6eXVsOUqGzWaIK+s0pp9TG6FM2/uuuXbmydrmUFaqg+b1O1byVJc5/o1eWaO7EBZpNoWYm7lf37V714wAtGjtW/+h+JoVQrWNYo7ewQOOo0hMyt2pXdqlQrRkZqitXrq3d76yK+pe9fX626rX9VRNFK1Lj9XFPvcsS1esGqH42qc5/QfW/vVUnd3cuY5Vy/e6ovzvbj6/Q2dnPOzF6bt8I8F+33H++6l6+XGdPyw3+X8vJPZ1tnd9F9aPpuunjA3WPvwvO09xnJyqguU+fXHedab117eITA8fls921j42tX3e9eweuM8989/2m+Wt/UOeWlPt/dbVCoWbEl2gRXVWn9VZ9b7zqNf1VE0TzE452rqSQMzSPsbqa8U5d861+HddDr2ew+7h6ShX0bXc9zJKlehdo3uuvN3CgBpefr5r3xmHNm/WY5sVP0Dc7TtGuFGomJVqU0Mu50f/QocZv8/Q7nPc06Teau6JEs7NVMzOqtWj6T+rmwaFDqrfdppqYqBXp6boatCI93UnAxu7bZ1uamen8qykzM+S2QrUjofaRn9Bf89Kmat6In+mqjmfrG5yhnSnU1JRyXbm8tjxz076n2cm7NFOaXp6N4tOu/GvFinrtiubkhDx2vbkfTr40cCVFeNvpHrANCbVO/rj/qXclRXZ2tWZmav220a3Btv/QIdVf3qiaGKfaOV711C6qnRNVE8Vp6/yu6GrpWz2aVQ/hxhEwXzKcMrimp+pnZ7V8DJEU7Jzqf+4NIOx+iGcfsy6uPUcEKIu8Z0fVzpuW61z9s/RmzTv/Us2LH6dvdpjgbme/bbhd8D8uAtQHlKtQ5fQ9enSoc37Uqn1OYsya5T43V7ap+8DDFqCcGuyLRPqYdR8fS0AX3XWXe59VOjvnBfc+e9TLqUU/H6C4n0mRk5zv9ONOzgqZUy3xbJEGcyBAfvn2d3Xq1ICfBcIS7LzV0HaakfNeQW75aLDd83smRURiCfb+3LeW6uzZqp98ojprVt3pzb0lt4mC9Z88x1LR2LG1cZ9zjvP7jBmq06e7c7OieeXTymjjt3t4nkkxwGfaU7TxZ1L4y8vL0+GgV058T6H2IWXeh5YtXKe6YUMrDVKgs4M8kC7QAyYbs69A69ezYUO9BW6/8mtvOXy4eHHdGN1l479uvffyw+W1y/nY9Nh7ge9t37BB9dFHVVev9tZPX5+HY05hvF4+Zb2C6uM3fKIJ8dUaF6c6adLXdff75sEGOkPonCl/cZZlvDPf535yLfyu6uunqT49Sm+/brl7/+N19rQnnHV+ebUzP8iJ11sOFzgPmFz9t1sDfzB4bYzqsoF11q93/K1+zanr5QOd/bnjqu1AUj8un+161q9Td08/HbzOFq5T/fCdgCe7Oh3WMa+7T4A9nIAv6Ke6c6vqaac5fx99tCrok+6Gf3rPf6qC3u6uz8dmP6nDQedM3+yth6KxYxs8nkNqiZOrX27U6VStWlV7fNfbN7p25crak3Bj9u03aKiq3s55sG011I6Es4+1K1fqU1B70s1ZU2eZJnfcmiKMdiWsYzecfAmUx/4dvnC2E6QNacy+G+q0h90Bn3Wx88Euf6Tq+omq5/eu8+GzNR+a2ax6CEewfJl5UW0nvKVjiKRQx77vuTeAsPsGvvvwnCPOf7huWQQqp6dvbXgQPFS74HtcvDamXn34Pgx79dVz650fvVatqu2jtPBDfKMiQDkF7Yu8Nibyx+yGDbrxvvsU0AceeKDuMXjbTap/H1Ivpx64Y7ACenT2hwqq/fm5fnjPdxrMqbD6qI3QYA4EyC9Pv9I3v/z7YmFpau42I+frCDRQEard+8X3Vd+cUrdeIhVLoPcX4LOGtx/l259qZcH6T57pHy5eXDvR/z34tkVNLZ9WFmqQIqHR12W0MlU9KCIrgbtE5EpgOHAucGp0I4u8/wD33FjJ3gzIynKm3Xqrc+lZ1qQTnWtK1qxplVhmztxGWVlfbxwe3niyAq/XkLDWHzGi/noP9WDLXqccSkv3143RUzZB1vUu9+OLa5fzMeSKU7ilAO6+G+bMgQkTfLbl2d6aNfwHgAWMGfMjDh1K4LUNJbxx7QH2dYQRs47jFynONlat6gGMr91v1xTnckfP5W9XHwOzjoIxLir/7Xyd5iUXL+OgDiJrdG8YXgH6LaROcvZd+QUc51wyd+vI53nng268/UEJM89fQtnhAWSdFu/c4tDApZUzL+9C2eH/JatzgdO38P0aQFUYchiSTq6zfv3jr9RZPuEoZ3/Hda/3dYKB43Jv9/P1AeruxOB15tlv9fFBvmt7M5ecu4ylK3/AnHfWMuEnX8L3h8L0lc57+MMfnO9hmjCBj+6/nwdvvJFLLtnG0qWTmMp4fnPtAbZ8DTWnHcd/lsI9V35Dwe4sbn27hIdn/oRmXfF5661w1VUwYQKVa9Y4sV6yjaVLxzBn4TomTAr5DdCB+R3fQ4Y4VyU6x+5EJnhm19s3VGZmwtKlzkHepUuT3odXt25hbStYOxLOPiozM/k+MIHNzJ78fzzx2ljm3LueCTed4Lz3bnDLLcLddzejPMMVTrviLueQx27Y+eKXx/HpdduQcLcToA1pzL7rHl91D4Gw+Nar5xLgrM3w4HCY3gcyk+ourxqZWzz8NbX8mhJHsHxZ9gzMvhRkcSPbzgiXRWOFOvZ9z72R2kd8irOPC16n7PBAssbEwdDOdffvKaeMA8759NJ0OONkKjdfDEwJv531PS6GuM8tdSxgzP+cRoekzmSdfQBGnBy4LiZO5NbROH2UJvaNYlqAcgrYFxkT5+5HRPiYHTGCvaWldSZ5j8Hz7oReu4Lm9RUXXcODT7zCFyXbOdA3tcGcam4ft9EC5NeF7n6lb37V64s1cdth5W6kct6/3Y3rXi+XAWaet4SysmyyzsiBCX+qWy8t0f4E2KbXxIlN2GDLCNZ/OjBgAIwb5/zh/15826JItc9RFD9v3rxox9Cg+fPnvwVMBxYBk4GfqepbodZZvHjxvLlz57ZGeBGTkZHBxIljmTs3g549nWmJiXDhhXj/9iw3btw4MjIyGtxeOMuFE4dHoHgas69g6zfEf71QMYb7Xnydcgp89hmceWbo2EaNGsm9957IRReVk5FxuM62TznFuQ25S5cyzj67kvPOO7F2v3HJkDIRDv8vZO2Co7p6T6YZnRKYOCaDuTP/l57DysBV6pxEeyyCtHOcdaoKQFJJTKxh3ClvkdGpuHadriUhT7wFBQUsWbKEK354JTffMoye6f/wbg+RkB8M/Mvds63LfvRH+vYoqrsd//cSJK6G6iTgfN/y8+wTIaNTAj+e/TVFu/ty5mmr6TmiN5z0VO176NkT+vUD4HCvXhzOyODHPx7Bl18m0md8FRNvvoS5czPYv/9zp4yumMO4cbtZsGQhl95wA3379g1+MDTEZ9+ecrv33mkcPnwUZ17Wi54jGpkEQQQ8dgPs+7LLLnPeT79+jUtAn23VE2Jb4eZeqH3s3buXFz/6iN/+caZTbnN61tmW971HsDwb4s2nK+Zw881967XNYR27jcgXr6ZsJ0Ab0th9N9Q2hmz7fevVP/6jukH3DrXLttQAhUek6qEhofKl/7HQt3n1EW2Nymua1g/JyMhg4pTrmHvBI/TM2hi6nFyl0H84dH+Ygq+KG9/OBjm3oMqoE9K495cfc91lL9CzT7eQddHUvk2b4VdOiYkd6vZFLv2Anq4tLXrMZmRkMGbMGHJycmqPwV5JAfO6qkr5svAQl1/UlemTl7NkxZtcdkEKffsNb/V6bEwOBMuv5sYV6rzV1JjCEqLdBXc+n1rJ3DlF9PzOggaPm2bF0oYEe5/hHEttrS2aP3/+1/PmzVscaF7MPzizqUaOHKnr1q2LdhjG1BXkgUJA8E56U9bxUVBQwJNPPln7IdV/e43oDNfZVnaXZsXVJIHKIgL78n1fQN3yioB6ddCKornv5orF2JscUzPzuFnbidS+I6GFcrhZ+/dorThiIYa2oJHl1Kz2ItrHZVsRq+UU4lgp2HaQJ5/fwmWXjKLvyCVHZD1G/Vwaq8eNibpQD860QQpjWltTGutIN/Ce7R1e73x7QFNPENE48TRjkMWYqIpUvsRCG9Ic0c7hWCiLWIihLWjNcor2cdlWxGo5WU7Ftlg9bkxUhRqkaMGbeI0xAXnu00sa7DTS4ZxEm7JOODFk/bR5J4hIx9XYfVbl20nOtB2RypdYaEOaI9o5HAtlEQsxtAWtWU7RPi7bilgtJ8up2Barx42JWXYlhTHR0pSrGSJ1BUSkRSOu6v2w/xlIvzg2ysCYcEX6Sqa22oZEO4djoSxiIYa2oDXLKdrHZVsRq+VkORXbYvW4MVFht3sYE6ua0ljHagMfq3EZE4silS/tqQ2Jhlgoi1iIoS2wcjLhsmPFmDbhiBykEJHdQGELbd4FFLfQtk10WJ22P1an7YvVZ/tjddq+WH22P1an7Y/VafvS1uvzKFXtGmhGux2kaEkisi7YqI9pm6xO2x+r0/bF6rP9sTptX6w+2x+r0/bH6rR9ac/1aQ/ONMYYY4wxxhhjTEywQQpjjDHGGGOMMcbEBBukaJrF0Q7ARJzVaftjddq+WH22P1an7YvVZ/tjddr+WJ22L+22Pu2ZFMYYY4wxxhhjjIkJdiWFMcYYY4wxxhhjYoINUhhjjDHGGGOMMSYm2CBFI4hIloi8KCIHRaRQRC6NdkwmOBFJFpHH3XW1X0T+IyJn+8yfICJbROSQiOSJyFF+6z4hIvtEZJeI3BCdd2GCEZEBInJYRJb5TLvUXd8HReQlEcnymWf5G8NE5GIR2eyun60icpp7uuVpGyMifUXkdRHZ466XP4lIgnvecBFZ767P9SIy3Gc9EZGFIvKt+7VQRCR67+TIJSLXiMg6ESkXkSf95jU5J0Ota1pOsPoUkVEiskpESkRkt4g8LyI9feaHzMlQ+WxaVqgc9VnmDhFREZnoM81yNAY10OamiMgjIlIsIqUi8i+fee02R22QonEeBiqA7sBM4FEROS66IZkQEoCvgLFAZ+B24Dl3B9oFrAR+BWQB64BnfdadBwwAjgLOAG4WkbNaL3QThoeBDz1/uHPxz8BsnBw9BDzit7zlbwwSkUnAQuAHQDpwOvCl5Wmb9QjwDdATGI7TBv9ERJKAl4FlQCawBHjZPR1gLnAecDwwDDgH+FHrhm7cdgJ3A0/4TmxOToaxrmk5AesTJw8XA31x6mw/8Fef+UFzMox8Ni0rWJ0CICLHADOAr/1mzcNyNBaFqs/FOPUx2P3zZz7z2m+Oqqq9wngBqTgfcAb6TFsK3BPt2OzVqHr8GLgAJ6nf86vfMmCQ+++dwJk+838NPBPt+O3lrY+LgedwTrbL3NMWAE/7LHOMO2fTLX9j+wW8B1wRYLrlaRt8AZuByT5//w5nAPFMYAfuh3a7520DzvI5Dub6zLsCeD/a7+dIfuF0mp/0+bvJOdnQuvZq/foMMP8EYL/P30FzsqF8tld06xR4A5gMFAATfaZbjsbwK0CbOwjYB3QKsny7zVG7kiJ8A4EqVf3cZ9pHgP0nto0Qke449bgJp94+8sxT1YPAVuA4EcnE+Q/gRz6rW13HCBHpBNwF+F/a71+nW3EPTGD5G7NEJB4YCXQVkS9EZLs4twd0xPK0rXoQuNh9iWpv4GycDvNxwMfq7im5fUxtndWpb6w+Y1FzcjLoui0cswnf6Th9JI9QOdlQPpsoEZEZQLmqvu433XK07TkJKATmu2/3+D8RucBnfrvNURukCF8azkiWr1Kc/9KaGCciicDfgCWqugWnPkv9FvPUZ5rP3/7zTPT9GnhcVbf7TW+oTi1/Y1N3IBGYDpyGc3vACJzbsyxP26Z/4XSC9gHbcS4ZfonQ9UmA+aVAmj2XIqY0Jycbqn8TRSIyDLgDuMlncqictPqMQSKSjnNl6fUBZluOtj19gKE49dALuAZYIiKD3fPbbY7aIEX4DgCd/KZ1wrl/z8QwEYnDubS/Aie5IXR9HvD523+eiSL3A38mAg8EmN1QnVr+xqYy988/qurXqloM/B7nMlXL0zbG3d6+gXNfcyrgwrkXdiEN56H//E7AAb//Apnoak5OWjsco0SkP/AP4HpVfcdnVqictPqMTfOApapaEGCe5WjbUwZUAneraoWqvg3k4dzKAe04R22QInyfAwkiMsBn2vHUvSzOxBj3SOLjOP+tvUBVK92zNuHUn2e5VJxnGGxS1T04Dxo63mdTVtexYRzOA762icgu4EbgAhHZQP067Qck4+Su5W+McufbdsD3g6jnd8vTticLyAH+pKrlqvotzoP4JuPUzTC/KyOGUVtndeobq89Y1JycDLpuC8dsQnB/e8Nq4NequtRvdqicbCifTXRMAK5zf3PHLiAb56Hxt1iOtkkfB5jm219qtzlqgxRhct+XtRK4S0RSRWQ0cC7Of+hN7HoU52m456hqmc/0F4GhInKBiHTAucTxY/etIABPAbeLSKaIDAJ+CDzZinGbwBbjnDCHu1+LgNeA7+LcznOOiJzmPrHeBaxU1f2WvzHvr8C1ItLNfc/sz4C/Y3na5rivhMkHfiwiCSKSAczB6WitAapxOtDJIuK5su0t98+ngBtEpLeI9AJ+jtVnVLjrrgMQD8SLSAdxvka2OTnZ0LqmhQSrT/czY97CGVRcFGDVUDm5htD5bFpQiBydgHN7gKeftBPn2x4edq9qORqDQtTnv3AedvlL9zKjcb6V5U33qu03R6P95M629ML5D9FLwEGcA+bSaMdkr5D1dRTOaONhnEuePK+Z7vkTgS04l1KtAfr6rJuM8zVA+4Ai4IZovx97Bazjebi/3cP996Xu3DyI87VLWT7zLH9j9IXzTIpHgL3ALuAhoIN7nuVpG3vhdIzXAHuAYpxv4ununjcCWO+uzw3ACJ/1BLgXKHG/7sXnqeT2atU6nOc+f/q+5rnnNTknQ61rr9avT+BO9+++faQDPuuFzMlQ+Wyv6NRpgOUKqPvtHpajMfhqoM09Dvg3Tv/1U+B8n/XabY6K+w0YY4wxxhhjjDHGRJXd7mGMMcYYY4wxxpiYYIMUxhhjjDHGGGOMiQk2SGGMMcYYY4wxxpiYYIMUxhhjjDHGGGOMiQk2SGGMMcYYY4wxxpiYYIMUxhhjjDHGGGOMiQk2SGGMMcaYViMi/xCRm5u7TIRjelJEHmut/RljjDEmuIRoB2CMMcaYI4eqnu37t4gocJqqrg22jDHGGGOOHHYlhTHGGGOMMcYYY2KCDVIYY4wxpg4RKRCRO0RkrYgcEJF1IvI/PvMT3PO/FJE9IpIrIkN95k8UkY0isk9EikVktc+8NSJyu/v3j9yT/+nez2P+y7j/HiYib7n39aWI3C4i8e55fUVERWS2iHwqIvtF5J8i0rORb7ujiCx1x7xVRC7z2f9lIvKFiNwiIl+LyDcicr+IJDZyH8YYY4xpgA1SGGOMMSaQq4DrgSxgBfC6iHRyz7sJ+D4wGegBvAOs8pn/FPAQ0BnoDdwdaAeqerz71zNVNU1Vr/RfRkQ6A6uAPPe+pgCXAzf4LXoRcLp7f6nAXY18vxcCb7rf74+AR0XkVJ/5RwE5QD/gFOAcnHIwxhhjTATZIIUxxhhjAnlcVderagWwECgDprrn/QBYqKpbVLUcZ0CgGmcAAaACOAborqrlqrqmGXFMcW/vbve2Nrvj8R/QmK+qxaq6D3gaGNnI/byvqstUtUpVVwMvAJf5zK8BblLVMlXdCtzrN98YY4wxEWCDFMYYY4wJpMDzi6oqsA3o456UDeT7zK9xL5/tnnQuMAD4P/ctGD9tRhzZQKE7Bo+tPvvy+Nrn94NAeiP3UxDg7z4+f3+jqodCzDfGGGNMBNi3exhjjDEmkL6eX0REcG512O6e9JXf/Dj3318BqOpHwEXu9cbgPHPiY1V9K8B+NMA0X18BR4mI+AxU9PPsK4L6Bvh7u8/f3UQkxWegwn++McYYYyLArqQwxhhjTCCXi8gJ7odD3gSkAK+55z0J3CwiA0UkCbgN5x8fr4lIkojMERGXe1BhD86tEtVB9rML56qLYF4DkoFb3ds+FrgFeDzcNyIi80SkoIHFRonIJSISLyLjgQuAJT7z44CFItJRRPoBN/rNN8YYY0wE2CCFMcYYYwJZjPPwyz04D6Wcoqql7nm/A5YD/wSKgPE4D7/c555/EbBFRA4ArwB3qurbQfZzG3CX+5s7/uw/073PM4GJ7n29ifNgzt834r3kAGsaWOY5nAeB7sEZALlaVd/1mV+Ic+VEPvAB8AbOcymMMcYYE0FS9xZPY4wxxhzp3Fcd3K6qy6IdSySIyOfABFVt0i0i7q8jvV1V+0c0MGOMMcbUY8+kMMYYY0y7pqoDox2DMcYYY8Jjt3sYY4wxxhhjjDEmJtjtHsYYY4wxxhhjjIkJdiWFMcYYY4wxxhhjYoINUhhjjDHGGGOMMSYm2CCFMcYYY4wxxhhjYoINUhhjjDHGGGOMMSYm2CCFMcYYY4wxxhhjYoINUhhjjDHGGGOMMSYm/D9IaH+AqudiowAAAABJRU5ErkJggg==\n"
          },
          "metadata": {
            "needs_background": "light"
          }
        }
      ],
      "source": [
        "fig, ax = plt.subplots(figsize=(18,3))\n",
        "\n",
        "ax.scatter(test.iloc[:, 1], test.iloc[:, 3], s=150, marker=\"D\", label='roommate', alpha=0.7, c='gold')\n",
        "ax.scatter(con1.iloc[:, 1], con1.iloc[:, 3], marker=\"1\", s=100, label='control #1', c='k')\n",
        "ax.scatter(con2.iloc[:, 1], con2.iloc[:, 3], marker=\"2\", s=100, label='control #2', c='r')\n",
        "ax.scatter(con3.iloc[:, 1], con3.iloc[:, 3], marker=\"3\", s=100, label='control #3', c='b')\n",
        "ax.set_ylim(-7, 107)\n",
        "ax.tick_params(axis='both', labelsize=12)\n",
        "plt.legend(fontsize=13)\n",
        "\n",
        "axins = ax.inset_axes([0.4, 0.4, 0.2, 0.2])\n",
        "left, right = 650, 700\n",
        "axins.scatter(test[(test['pos'] > left) & (test['pos'] < right)].iloc[:, 1], \n",
        "              test[(test['pos'] > left) & (test['pos'] < right)].iloc[:, 3], s=150, marker=\"D\", alpha=0.7, c='gold')\n",
        "axins.scatter(con1[(con1['pos'] > left) & (con1['pos'] < right)].iloc[:, 1], \n",
        "              con1[(con1['pos'] > left) & (con1['pos'] < right)].iloc[:, 3], marker=\"1\", s=100, c='k')\n",
        "axins.scatter(con2[(con2['pos'] > left) & (con2['pos'] < right)].iloc[:, 1], \n",
        "              con2[(con2['pos'] > left) & (con2['pos'] < right)].iloc[:, 3], marker=\"2\", s=100, c='r')\n",
        "axins.scatter(con3[(con3['pos'] > left) & (con3['pos'] < right)].iloc[:, 1], \n",
        "              con3[(con3['pos'] > left) & (con3['pos'] < right)].iloc[:, 3], marker=\"3\", s=100, c='b')\n",
        "\n",
        "x1, x2, y1, y2 = left, right, -1, 3\n",
        "ax.indicate_inset_zoom(axins, edgecolor=\"gray\")\n",
        "axins.set_xlim(x1, x2)\n",
        "axins.set_ylim(y1, y2)\n",
        "axins.tick_params(axis='both', labelsize=11)\n",
        "\n",
        "plt.xlabel('position, bp', fontsize=13)\n",
        "plt.ylabel('Variant coverage, %', fontsize=13)\n",
        "plt.title('Mutation frequency in our roommate and controls', fontsize=15)\n",
        "# plt.savefig('mut_frequensy.pdf', bbox_inches='tight')"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Optional Extra-Credit Challenge Question"
      ],
      "metadata": {
        "id": "-PE20tMDiQpI"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "1. How would you calculate the ACTUAL average coverage per position for one of our data sets, only for mapped reads, and taking into consideration the fact that the reads can be not all the same length?\n"
      ],
      "metadata": {
        "id": "PvpjkZkmiSHP"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!bedtools bamtobed -i roommate_sorted.bam > roommate_sorted.bed\n",
        "!bedtools genomecov -d -i roommate_sorted.bed -g reference.fasta.fai > coverage.txt\n",
        "!head coverage.txt -n 10"
      ],
      "metadata": {
        "id": "9Ekzci-Xh-ZD",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "95d8c9e8-e824-4ac9-999a-f0706e247463"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "KF848938.1\t1\t14421\n",
            "KF848938.1\t2\t14447\n",
            "KF848938.1\t3\t14580\n",
            "KF848938.1\t4\t14621\n",
            "KF848938.1\t5\t14641\n",
            "KF848938.1\t6\t14686\n",
            "KF848938.1\t7\t14783\n",
            "KF848938.1\t8\t14798\n",
            "KF848938.1\t9\t15055\n",
            "KF848938.1\t10\t15074\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "pd.read_table('coverage.txt', header=None)[2].max()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "_1sxwacRrm_V",
        "outputId": "e1c4d17d-387b-4050-b9ef-55721a59c42d"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "44523"
            ]
          },
          "metadata": {},
          "execution_count": 64
        }
      ]
    }
  ],
  "metadata": {
    "colab": {
      "collapsed_sections": [
        "cPwkCuvYuZKw",
        "_KPOdLQDXIvb",
        "JNPwpi3SXOEJ",
        "S_ezD11FudN0",
        "WhZjddiY7TjA",
        "ryFvNRbt7age"
      ],
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}